// radixdlt/build/main/modules/txbuilder/RadixTransactionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var rxjs_1 = require("rxjs");
var typescript_map_1 = require("typescript-map");
var decimal_js_1 = tslib_1.__importDefault(require("decimal.js"));
var bn_js_1 = tslib_1.__importDefault(require("bn.js"));
var __1 = require("../..");
var atommodel_1 = require("../atommodel");
var RadixLogger_1 = require("../common/RadixLogger");
var RadixTokenDefinition_1 = require("../token/RadixTokenDefinition");
var RadixTransactionBuilder = /** @class */ (function () {
    function RadixTransactionBuilder() {
        this.BNZERO = new bn_js_1.default(0);
        this.DCZERO = new decimal_js_1.default(0);
        this.participants = new typescript_map_1.TSMap();
        this.particleGroups = [];
    }
    RadixTransactionBuilder.prototype.getSubUnitsQuantity = function (decimalQuantity) {
        if (typeof decimalQuantity !== 'number' && typeof decimalQuantity !== 'string' && !decimal_js_1.default.isDecimal(decimalQuantity)) {
            throw new Error('quantity is not a valid number');
        }
        var unitsQuantity = new decimal_js_1.default(decimalQuantity);
        var subunitsQuantity = RadixTokenDefinition_1.RadixTokenDefinition.fromDecimalToSubunits(unitsQuantity);
        return subunitsQuantity;
    };
    /**
     * Creates transfer atom
     * @param from Sender account, needs to have RadixAccountTransferSystem
     * @param to Receiver account
     * @param tokenReference TokenClassReference string
     * @param decimalQuantity
     * @param [message] Optional reference message
     */
    RadixTransactionBuilder.createTransferAtom = function (from, to, tokenReference, decimalQuantity, message) {
        return new RadixTransactionBuilder().addTransfer(from, to, tokenReference, decimalQuantity, message);
    };
    /**
     * Creates transfer atom
     * @param from Sender account, needs to have RadixAccountTransferSystem
     * @param to Receiver account
     * @param tokenReferenceURI TokenClassReference string
     * @param decimalQuantity
     * @param [message] Optional reference message
     */
    RadixTransactionBuilder.prototype.addTransfer = function (from, to, tokenReference, decimalQuantity, message) {
        var e_1, _a;
        tokenReference = (tokenReference instanceof atommodel_1.RRI)
            ? tokenReference
            : atommodel_1.RRI.fromString(tokenReference);
        var subunitsQuantity = this.getSubUnitsQuantity(decimalQuantity);
        if (subunitsQuantity.lt(this.BNZERO)) {
            throw new Error('Negative quantity is not allowed');
        }
        else if (subunitsQuantity.eq(this.BNZERO)) {
            throw new Error("Quantity 0 is not valid");
        }
        var transferSytem = from.transferSystem;
        if (subunitsQuantity.gt(transferSytem.balance[tokenReference.toString()])) {
            throw new Error('Insufficient funds');
        }
        var unspentConsumables = transferSytem.getUnspentConsumables();
        var createTransferAtomParticleGroup = new __1.RadixParticleGroup();
        var consumerQuantity = new bn_js_1.default(0);
        var granularity = new bn_js_1.default(1);
        var tokenPermissions;
        try {
            for (var unspentConsumables_1 = tslib_1.__values(unspentConsumables), unspentConsumables_1_1 = unspentConsumables_1.next(); !unspentConsumables_1_1.done; unspentConsumables_1_1 = unspentConsumables_1.next()) {
                var consumable = unspentConsumables_1_1.value;
                if (!consumable.getTokenDefinitionReference().equals(tokenReference)) {
                    continue;
                }
                // Assumes all consumables of a token have the same granularity and permissions(enforced by core)
                granularity = consumable.getGranularity();
                tokenPermissions = consumable.getTokenPermissions();
                createTransferAtomParticleGroup.particles.push(atommodel_1.RadixSpunParticle.down(consumable));
                consumerQuantity.iadd(consumable.getAmount());
                if (consumerQuantity.gte(subunitsQuantity)) {
                    break;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (unspentConsumables_1_1 && !unspentConsumables_1_1.done && (_a = unspentConsumables_1.return)) _a.call(unspentConsumables_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        createTransferAtomParticleGroup.particles.push(atommodel_1.RadixSpunParticle.up(new atommodel_1.RadixTransferrableTokensParticle(subunitsQuantity, granularity, to.address, Date.now(), tokenReference, tokenPermissions)));
        // Remainder to myself
        if (consumerQuantity.sub(subunitsQuantity).gtn(0)) {
            createTransferAtomParticleGroup.particles.push(atommodel_1.RadixSpunParticle.up(new atommodel_1.RadixTransferrableTokensParticle(consumerQuantity.sub(subunitsQuantity), granularity, from.address, Date.now(), tokenReference, tokenPermissions)));
        }
        if (!subunitsQuantity.mod(granularity).eq(this.BNZERO)) {
            throw new Error("This token requires that any tranferred amount is a multiple of it's granularity = \n                " + RadixTokenDefinition_1.RadixTokenDefinition.fromSubunitsToDecimal(granularity));
        }
        this.participants.set(from.getAddress(), from);
        this.participants.set(to.getAddress(), to);
        if (message) {
            this.addEncryptedMessage(from, 'transfer', message, [to, from]);
        }
        this.particleGroups.push(createTransferAtomParticleGroup);
        return this;
    };
    /**
     * Create an atom to burn a specified amount of tokens
     * ownerAccount must be the owner and the holder of the tokens to be burned
     *
     * @param  {RadixAccount} ownerAccount
     * @param  {string|RRI} tokenReference
     * @param  {string|number|Decimal} decimalQuantity
     */
    RadixTransactionBuilder.createBurnAtom = function (ownerAccount, tokenReference, decimalQuantity) {
        return new this().burnTokens(ownerAccount, tokenReference, decimalQuantity);
    };
    /**
     * Create an atom to burn a specified amount of tokens
     * The token must be multi-issuance
     *
     * @param  {RadixAccount} ownerAccount must be the owner and the holder of the tokens to be burned
     * @param  {string|RRI} tokenReference
     * @param  {string|number|Decimal} decimalQuantity
     */
    RadixTransactionBuilder.prototype.burnTokens = function (ownerAccount, tokenReference, decimalQuantity) {
        var e_2, _a;
        tokenReference = (tokenReference instanceof atommodel_1.RRI)
            ? tokenReference
            : atommodel_1.RRI.fromString(tokenReference);
        var tokenClass = ownerAccount.tokenDefinitionSystem.getTokenDefinition(tokenReference.getName());
        var subunitsQuantity = this.getSubUnitsQuantity(decimalQuantity);
        if (subunitsQuantity.lt(this.BNZERO)) {
            throw new Error('Negative quantity is not allowed');
        }
        else if (subunitsQuantity.eq(this.BNZERO)) {
            throw new Error("Quantity 0 is not valid");
        }
        var transferSytem = ownerAccount.transferSystem;
        if (tokenClass.tokenSupplyType !== RadixTokenDefinition_1.RadixTokenSupplyType.MUTABLE) {
            throw new Error('This token is fixed supply');
        }
        if (subunitsQuantity.gt(transferSytem.balance[tokenReference.toString()])) {
            throw new Error('Insufficient funds');
        }
        if (!subunitsQuantity.mod(tokenClass.getGranularity()).eq(this.BNZERO)) {
            throw new Error("This token requires that any tranferred amount is a multiple of it's granularity = \n                " + RadixTokenDefinition_1.RadixTokenDefinition.fromSubunitsToDecimal(tokenClass.getGranularity()));
        }
        var unspentConsumables = transferSytem.getUnspentConsumables();
        var burnParticleGroup = new __1.RadixParticleGroup();
        var consumerQuantity = new bn_js_1.default(0);
        var tokenPermissions;
        try {
            for (var unspentConsumables_2 = tslib_1.__values(unspentConsumables), unspentConsumables_2_1 = unspentConsumables_2.next(); !unspentConsumables_2_1.done; unspentConsumables_2_1 = unspentConsumables_2.next()) {
                var consumable = unspentConsumables_2_1.value;
                if (!consumable.getTokenDefinitionReference().equals(tokenReference)) {
                    continue;
                }
                burnParticleGroup.particles.push(atommodel_1.RadixSpunParticle.down(consumable));
                tokenPermissions = consumable.getTokenPermissions();
                consumerQuantity.iadd(consumable.getAmount());
                if (consumerQuantity.gte(subunitsQuantity)) {
                    break;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (unspentConsumables_2_1 && !unspentConsumables_2_1.done && (_a = unspentConsumables_2.return)) _a.call(unspentConsumables_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
        burnParticleGroup.particles.push(atommodel_1.RadixSpunParticle.up(new atommodel_1.RadixUnallocatedTokensParticle(subunitsQuantity, tokenClass.getGranularity(), Date.now(), tokenReference, tokenPermissions)));
        // Remainder to myself
        if (consumerQuantity.sub(subunitsQuantity).gtn(0)) {
            burnParticleGroup.particles.push(atommodel_1.RadixSpunParticle.up(new atommodel_1.RadixTransferrableTokensParticle(consumerQuantity.sub(subunitsQuantity), tokenClass.getGranularity(), ownerAccount.address, Date.now(), tokenReference, tokenPermissions)));
        }
        this.particleGroups.push(burnParticleGroup);
        this.participants.set(ownerAccount.getAddress(), ownerAccount);
        return this;
    };
    /**
     * Create an atom to mint a specified amount of tokens
     * The token must be multi-issuance
     *
     * @param  {RadixAccount} ownerAccount must be the owner of the token
     * @param  {string|RRI} tokenReference
     * @param  {string|number|Decimal} decimalQuantity
     */
    RadixTransactionBuilder.createMintAtom = function (ownerAccount, tokenReference, decimalQuantity) {
        return new this().mintTokens(ownerAccount, tokenReference, decimalQuantity);
    };
    /**
     * Create an atom to mint a specified amount of tokens
     * The token must be multi-issuance
     *
     * @param  {RadixAccount} ownerAccount must be the owner of the token
     * @param  {string|RRI} tokenReference
     * @param  {string|number|Decimal} decimalQuantity
     */
    RadixTransactionBuilder.prototype.mintTokens = function (ownerAccount, tokenReference, decimalQuantity) {
        var e_3, _a;
        tokenReference = (tokenReference instanceof atommodel_1.RRI)
            ? tokenReference
            : atommodel_1.RRI.fromString(tokenReference);
        var tokenClass = ownerAccount.tokenDefinitionSystem.getTokenDefinition(tokenReference.getName());
        var subunitsQuantity = this.getSubUnitsQuantity(decimalQuantity);
        if (subunitsQuantity.lt(this.BNZERO)) {
            throw new Error('Negative quantity is not allowed');
        }
        else if (subunitsQuantity.eq(this.BNZERO)) {
            throw new Error("Quantity 0 is not valid");
        }
        if (tokenClass.tokenSupplyType !== RadixTokenDefinition_1.RadixTokenSupplyType.MUTABLE) {
            throw new Error('This token is fixed supply');
        }
        if (subunitsQuantity.gte(tokenClass.getUnallocatedSupply())) {
            throw new Error('Total supply would exceed 2^256');
        }
        if (!subunitsQuantity.mod(tokenClass.getGranularity()).eq(this.BNZERO)) {
            throw new Error("This token requires that any tranferred amount is a multiple of it's granularity = \n                " + RadixTokenDefinition_1.RadixTokenDefinition.fromSubunitsToDecimal(tokenClass.getGranularity()));
        }
        this.participants.set(ownerAccount.getAddress(), ownerAccount);
        var unallocatedTokens = tokenClass.getUnallocatedTokens();
        var tokenPermissions = unallocatedTokens[0].getTokenPermissions();
        var particleGroup = new __1.RadixParticleGroup();
        var consumerQuantity = new bn_js_1.default(0);
        try {
            for (var unallocatedTokens_1 = tslib_1.__values(unallocatedTokens), unallocatedTokens_1_1 = unallocatedTokens_1.next(); !unallocatedTokens_1_1.done; unallocatedTokens_1_1 = unallocatedTokens_1.next()) {
                var consumable = unallocatedTokens_1_1.value;
                particleGroup.particles.push(atommodel_1.RadixSpunParticle.down(consumable));
                consumerQuantity.iadd(consumable.getAmount());
                if (consumerQuantity.gte(subunitsQuantity)) {
                    break;
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (unallocatedTokens_1_1 && !unallocatedTokens_1_1.done && (_a = unallocatedTokens_1.return)) _a.call(unallocatedTokens_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
        // Remainder
        if (consumerQuantity.sub(subunitsQuantity).gtn(0)) {
            particleGroup.particles.push(atommodel_1.RadixSpunParticle.up(new atommodel_1.RadixUnallocatedTokensParticle(consumerQuantity.sub(subunitsQuantity), tokenClass.getGranularity(), Date.now(), tokenReference, tokenPermissions)));
        }
        var particle = new atommodel_1.RadixTransferrableTokensParticle(subunitsQuantity, tokenClass.getGranularity(), ownerAccount.address, Date.now(), tokenReference, tokenPermissions);
        particleGroup.particles.push(atommodel_1.RadixSpunParticle.up(particle));
        this.particleGroups.push(particleGroup);
        return this;
    };
    RadixTransactionBuilder.prototype.createToken = function (owner, name, symbol, description, granularity, decimalQuantity, iconUrl, permissions) {
        var tokenAmount = this.getSubUnitsQuantity(decimalQuantity);
        var tokenGranularity = this.getSubUnitsQuantity(granularity);
        if (tokenAmount.lt(this.BNZERO)) {
            throw new Error('Negative quantity is not allowed');
        }
        this.participants.set(owner.getAddress(), owner);
        var tokenClassParticle = new atommodel_1.RadixTokenDefinitionParticle(owner.address, name, symbol, description, tokenGranularity, iconUrl, permissions);
        var rriParticle = new atommodel_1.RadixRRIParticle(tokenClassParticle.getRRI());
        var initialSupplyParticle = new atommodel_1.RadixUnallocatedTokensParticle(new bn_js_1.default(2).pow(new bn_js_1.default(256)).subn(1), tokenGranularity, Date.now(), tokenClassParticle.getTokenDefinitionReference(), permissions);
        var createTokenParticleGroup = new __1.RadixParticleGroup([
            atommodel_1.RadixSpunParticle.down(rriParticle),
            atommodel_1.RadixSpunParticle.up(tokenClassParticle),
            atommodel_1.RadixSpunParticle.up(initialSupplyParticle),
        ]);
        this.particleGroups.push(createTokenParticleGroup);
        if (tokenAmount.gtn(0)) {
            var mintParticle = new atommodel_1.RadixTransferrableTokensParticle(tokenAmount, tokenGranularity, owner.address, Date.now(), tokenClassParticle.getTokenDefinitionReference(), permissions);
            var mintParticleGroup = new __1.RadixParticleGroup([
                atommodel_1.RadixSpunParticle.down(initialSupplyParticle),
                atommodel_1.RadixSpunParticle.up(mintParticle),
            ]);
            var remainder = initialSupplyParticle.getAmount().sub(mintParticle.getAmount());
            if (remainder.gten(0)) {
                // Remainder
                var remainingSupplyParticle = new atommodel_1.RadixUnallocatedTokensParticle(remainder, tokenGranularity, Date.now(), tokenClassParticle.getTokenDefinitionReference(), permissions);
                mintParticleGroup.particles.push(atommodel_1.RadixSpunParticle.up(remainingSupplyParticle));
            }
            this.particleGroups.push(mintParticleGroup);
        }
        return this;
    };
    RadixTransactionBuilder.prototype.createTokenSingleIssuance = function (owner, name, symbol, description, granularity, amount, iconUrl) {
        if (granularity === void 0) { granularity = new decimal_js_1.default('1e-18'); }
        var permissions = {
            mint: atommodel_1.RadixTokenPermissionsValues.TOKEN_CREATION_ONLY,
            burn: atommodel_1.RadixTokenPermissionsValues.TOKEN_CREATION_ONLY,
        };
        if (new decimal_js_1.default(amount).eq(0)) {
            throw new Error('Single-issuance tokens cannot have an amount of 0');
        }
        return this.createToken(owner, name, symbol, description, granularity, amount, iconUrl, permissions);
    };
    RadixTransactionBuilder.prototype.createTokenMultiIssuance = function (owner, name, symbol, description, granularity, amount, iconUrl) {
        if (granularity === void 0) { granularity = new decimal_js_1.default('1e-18'); }
        var permissions = {
            mint: atommodel_1.RadixTokenPermissionsValues.TOKEN_OWNER_ONLY,
            burn: atommodel_1.RadixTokenPermissionsValues.TOKEN_OWNER_ONLY,
        };
        return this.createToken(owner, name, symbol, description, granularity, amount, iconUrl, permissions);
    };
    /**
     * Creates payload atom
     * @param from
     * @param recipients Everyone who will receive and be able to decrypt the message
     * @param applicationId
     * @param payload
     * @param [encrypted] Sets if the message should be encrypted using ECIES
     */
    RadixTransactionBuilder.createPayloadAtom = function (from, recipients, applicationId, payload, encrypted) {
        if (encrypted === void 0) { encrypted = true; }
        if (encrypted) {
            return new RadixTransactionBuilder().addEncryptedMessage(from, applicationId, payload, recipients);
        }
        else {
            return new RadixTransactionBuilder().addUnencryptedMessage(from, applicationId, payload, recipients);
        }
    };
    /**
     * Creates radix messaging application payload atom
     * @param from
     * @param to
     * @param message
     */
    RadixTransactionBuilder.createRadixMessageAtom = function (from, to, message) {
        return new RadixTransactionBuilder().addEncryptedMessage(from, 'message', message, [from, to]);
    };
    RadixTransactionBuilder.prototype.addEncryptedMessage = function (from, applicationId, message, recipients) {
        var recipientPubKeys = recipients.map(function (r) { return r.address.getPublic(); });
        var _a = __1.RadixECIES.encryptForMultiple(recipientPubKeys, Buffer.from(message)), protectors = _a.protectors, ciphertext = _a.ciphertext;
        this.addMessageParticle(from, ciphertext, {
            application: applicationId,
        }, recipients);
        this.addMessageParticle(from, JSON.stringify(protectors.map(function (p) { return p.toString('base64'); })), {
            application: 'encryptor',
            contentType: 'application/json',
        }, recipients);
        return this;
    };
    RadixTransactionBuilder.prototype.addUnencryptedMessage = function (from, applicationId, message, recipients) {
        this.addMessageParticle(from, message, {
            application: applicationId,
        }, recipients);
        return this;
    };
    RadixTransactionBuilder.prototype.addMessageParticle = function (from, data, metadata, recipients) {
        var e_4, _a;
        try {
            for (var recipients_1 = tslib_1.__values(recipients), recipients_1_1 = recipients_1.next(); !recipients_1_1.done; recipients_1_1 = recipients_1.next()) {
                var recipient = recipients_1_1.value;
                this.participants.set(recipient.getAddress(), recipient);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (recipients_1_1 && !recipients_1_1.done && (_a = recipients_1.return)) _a.call(recipients_1);
            }
            finally { if (e_4) throw e_4.error; }
        }
        var particle = new atommodel_1.RadixMessageParticle(from.address, (recipients.length === 1) ? recipients[0].address : recipients[1].address, data, metadata);
        var particleParticleGroup = new __1.RadixParticleGroup([atommodel_1.RadixSpunParticle.up(particle)]);
        this.particleGroups.push(particleParticleGroup);
        return this;
    };
    /**
     * Add a particle which enforces that only one such particle can exists per account,
     * enforced on the ledger level
     *
     * @param  {RadixAccount} account Scope of the uniqueness constraint
     * @param  {string} unique The unique id
     */
    RadixTransactionBuilder.prototype.addUniqueParticle = function (account, unique) {
        var uniqueParticle = new atommodel_1.RadixUniqueParticle(account.address, unique);
        var rriParticle = new atommodel_1.RadixRRIParticle(uniqueParticle.getRRI());
        var uniqueParticleGroup = new __1.RadixParticleGroup([
            atommodel_1.RadixSpunParticle.down(rriParticle),
            atommodel_1.RadixSpunParticle.up(uniqueParticle),
        ]);
        this.particleGroups.push(uniqueParticleGroup);
        return this;
    };
    /**
     * Builds the atom, finds a node to submit to, adds network fee, signs the atom and submits
     * @param signer
     * @returns a BehaviourSubject that streams the atom status updates
     */
    RadixTransactionBuilder.prototype.signAndSubmit = function (signer) {
        var _this = this;
        var atom = this.buildAtom();
        var stateSubject = new rxjs_1.BehaviorSubject('FINDING_NODE');
        // Find a shard, any of the participant shards is ok
        var shard = atom.getShards()[0];
        // Get node from universe
        __1.radixUniverse.getNodeConnection(shard)
            .then(function (connection) {
            RadixTransactionBuilder.signAndSubmitAtom(atom, connection, signer, _this.participants.values())
                .subscribe(stateSubject);
        });
        return stateSubject;
    };
    RadixTransactionBuilder.prototype.buildAtom = function () {
        if (this.particleGroups.length === 0) {
            throw new Error('No particle groups specified');
        }
        var atom = new atommodel_1.RadixAtom();
        atom.particleGroups = this.particleGroups;
        // Add timestamp
        atom.setTimestamp(Date.now());
        return atom;
    };
    RadixTransactionBuilder.signAndSubmitAtom = function (atom, connection, signer, participants) {
        var signedAtom = null;
        // Add POW fee
        var stateSubject = new rxjs_1.BehaviorSubject('GENERATING_POW');
        atom.clearPowNonce();
        __1.RadixFeeProvider.generatePOWFee(__1.radixUniverse.universeConfig.getMagic(), atom).then(function (pow) {
            atom.setPowNonce(pow.nonce);
            // Sign atom
            stateSubject.next('SIGNING');
            return signer.signAtom(atom);
        }).then(function (_signedAtom) {
            var e_5, _a;
            signedAtom = _signedAtom;
            try {
                // Push atom into participant accounts to minimize delay
                for (var participants_1 = tslib_1.__values(participants), participants_1_1 = participants_1.next(); !participants_1_1.done; participants_1_1 = participants_1.next()) {
                    var participant = participants_1_1.value;
                    participant._onAtomReceived({
                        action: 'STORE',
                        atom: signedAtom,
                        processedData: {},
                        isHead: true,
                    });
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (participants_1_1 && !participants_1_1.done && (_a = participants_1.return)) _a.call(participants_1);
                }
                finally { if (e_5) throw e_5.error; }
            }
            var submissionSubject = connection.submitAtom(signedAtom);
            submissionSubject.subscribe(stateSubject);
            submissionSubject.subscribe({
                error: function (error) {
                    var e_6, _a;
                    RadixLogger_1.logger.info('Problem submitting atom, deleting', error);
                    try {
                        // Delete atom from participant accounts
                        for (var participants_2 = tslib_1.__values(participants), participants_2_1 = participants_2.next(); !participants_2_1.done; participants_2_1 = participants_2.next()) {
                            var participant = participants_2_1.value;
                            participant._onAtomReceived({
                                action: 'DELETE',
                                atom: signedAtom,
                                processedData: {},
                                isHead: true,
                            });
                        }
                    }
                    catch (e_6_1) { e_6 = { error: e_6_1 }; }
                    finally {
                        try {
                            if (participants_2_1 && !participants_2_1.done && (_a = participants_2.return)) _a.call(participants_2);
                        }
                        finally { if (e_6) throw e_6.error; }
                    }
                }
            });
        }).catch(function (error) {
            stateSubject.error(error);
        });
        return stateSubject;
    };
    return RadixTransactionBuilder;
}());
exports.default = RadixTransactionBuilder;
