// @radixdlt/application/dist/dto/_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionStatus = exports.TransactionType = exports.TransactionTrackingEventType = void 0;
var TransactionTrackingEventType;
(function (TransactionTrackingEventType) {
    /* A TransactionIntent was successfully created and any message has been encrypted */
    TransactionTrackingEventType["INITIATED"] = "INITIATED";
    TransactionTrackingEventType["BUILT_FROM_INTENT"] = "BUILT_FROM_INTENT";
    TransactionTrackingEventType["SIGNED"] = "SIGNED";
    TransactionTrackingEventType["SUBMITTED"] = "SUBMITTED";
    TransactionTrackingEventType["ASKED_FOR_CONFIRMATION"] = "ASKED_FOR_CONFIRMATION";
    TransactionTrackingEventType["CONFIRMED"] = "CONFIRMED";
    /* API has finished "finalizing" / "confirming" the transaction, which now is pending. */
    TransactionTrackingEventType["FINALIZED"] = "FINALIZED";
    TransactionTrackingEventType["UPDATE_OF_STATUS_OF_PENDING_TX"] = "UPDATE_OF_STATUS_OF_PENDING_TX";
    TransactionTrackingEventType["COMPLETED"] = "COMPLETED";
})(TransactionTrackingEventType = exports.TransactionTrackingEventType || (exports.TransactionTrackingEventType = {}));
var TransactionType;
(function (TransactionType) {
    TransactionType["FROM_ME_TO_ME"] = "FROM_ME_TO_ME";
    TransactionType["INCOMING"] = "INCOMING";
    TransactionType["OUTGOING"] = "OUTGOING";
    TransactionType["UNRELATED"] = "UNRELATED";
})(TransactionType = exports.TransactionType || (exports.TransactionType = {}));
var TransactionStatus;
(function (TransactionStatus) {
    TransactionStatus["PENDING"] = "PENDING";
    TransactionStatus["CONFIRMED"] = "CONFIRMED";
    TransactionStatus["FAILED"] = "FAILED";
})(TransactionStatus = exports.TransactionStatus || (exports.TransactionStatus = {}));
