// radixdlt/build/module/modules/account/RadixDecryptionAccountSystem.js
import { RadixMessageParticle } from '../atommodel';
export var RadixDecryptionState;
(function (RadixDecryptionState) {
    RadixDecryptionState["DECRYPTED"] = "DECRYPTED";
    RadixDecryptionState["NOT_ENCRYPTED"] = "NOT_ENCRYPTED";
    RadixDecryptionState["CANNOT_DECRYPT"] = "CANNOT_DECRYPT";
})(RadixDecryptionState || (RadixDecryptionState = {}));
export class RadixDecryptionAccountSystem {
    constructor(decryptionProvider) {
        this.name = 'DECRYPTION';
        if (decryptionProvider) {
            this.decryptionProvider = decryptionProvider;
        }
    }
    async processAtomUpdate(atomUpdate) {
        const atom = atomUpdate.atom;
        const messageParticles = atom.getParticlesOfType(RadixMessageParticle);
        const dataParticle = messageParticles.find(p => {
            return p.getMetaData('application') !== 'encryptor';
        });
        const encryptorParticle = messageParticles.find(p => {
            return p.getMetaData('application') === 'encryptor';
        });
        if (!(dataParticle)) {
            return;
        }
        let decryptedData;
        if (encryptorParticle) {
            const protectors = encryptorParticle.getData().asJSON().map(p => Buffer.from(p, 'base64'));
            const payload = dataParticle.getData().bytes;
            try {
                const decrypted = await this.decryptionProvider.decryptECIESPayloadWithProtectors(protectors, payload);
                decryptedData = {
                    data: decrypted.toString(),
                    decryptionState: RadixDecryptionState.DECRYPTED,
                    application: dataParticle.getMetaData('application'),
                    from: dataParticle.from,
                    to: dataParticle.getAddresses(),
                };
            }
            catch (error) {
                decryptedData = {
                    data: dataParticle.getData().toString(),
                    decryptionState: RadixDecryptionState.CANNOT_DECRYPT,
                    application: dataParticle.getMetaData('application'),
                    from: dataParticle.from,
                    to: dataParticle.getAddresses(),
                };
            }
        }
        else {
            decryptedData = {
                data: dataParticle.getData().bytes.toString(),
                decryptionState: RadixDecryptionState.NOT_ENCRYPTED,
                application: dataParticle.getMetaData('application'),
                from: dataParticle.from,
                to: dataParticle.getAddresses(),
            };
        }
        atomUpdate.processedData.decryptedData = decryptedData;
    }
}
