// radixdlt/build/module/modules/account/RadixAccount.js
import { BehaviorSubject } from 'rxjs';
import { TSMap } from 'typescript-map';
import { RadixTransferAccountSystem, RadixMessagingAccountSystem, RadixDecryptionAccountSystem, RadixDataAccountSystem, RadixCacheAccountSystem, radixUniverse, RadixTokenDefinitionAccountSystem, } from '../..';
import { logger } from '../common/RadixLogger';
import { RadixAddress } from '../atommodel';
import { radixHash } from '../common/RadixUtil';
export default class RadixAccount {
    /**
     * Creates an instance of radix account.
     * @param address Address of the account
     * @param [plain] If set to false, will not create default account systems.
     * Use this for accounts that will not be connected to the network
     */
    constructor(address, plain = false) {
        this.address = address;
        this.accountSystems = new TSMap();
        this.synced = new BehaviorSubject(false);
        this.connectionStatus = new BehaviorSubject('STARTING');
        this.openNodeConnection = async () => {
            this.connectionStatus.next('CONNECTING');
            try {
                this.nodeConnection = await radixUniverse.getNodeConnection(this.address.getShard());
                this.nodeConnection.on('closed', this._onConnectionClosed);
                // Subscribe to events
                this.atomSubscription = this.nodeConnection.subscribe(this.address.toString());
                this.atomSubscription.subscribe({
                    next: this._onAtomReceived,
                    error: error => logger.error('Subscription error:', error),
                });
                this.connectionStatus.next('CONNECTED');
            }
            catch (error) {
                logger.error(error);
                setTimeout(this._onConnectionClosed, 1000);
            }
        };
        /**
         * Unsubscribes the node connection to the stream of past and future atoms associated with this address account
         *
         * @returns A promise with the result of the unsubscription call
         */
        this.closeNodeConnection = async () => {
            this.connectionStatus.next('DISCONNECTED');
            if (this.nodeConnection) {
                this.nodeConnection.removeListener('closed', this._onConnectionClosed);
                return this.nodeConnection.unsubscribe(this.getAddress());
            }
        };
        this._onAtomReceived = async (atomUpdate) => {
            for (const system of this.accountSystems.values()) {
                await system.processAtomUpdate(atomUpdate);
            }
            this.synced.next(atomUpdate.isHead);
        };
        this._onConnectionClosed = () => {
            // Get a new one
            this.openNodeConnection();
        };
        if (!plain) {
            this.cacheSystem = new RadixCacheAccountSystem(address);
            this.addAccountSystem(this.cacheSystem);
            this.decryptionSystem = new RadixDecryptionAccountSystem();
            this.addAccountSystem(this.decryptionSystem);
            this.tokenDefinitionSystem = new RadixTokenDefinitionAccountSystem(address);
            this.addAccountSystem(this.tokenDefinitionSystem);
            this.transferSystem = new RadixTransferAccountSystem(address);
            this.addAccountSystem(this.transferSystem);
            this.dataSystem = new RadixDataAccountSystem(address);
            this.addAccountSystem(this.dataSystem);
            this.messagingSystem = new RadixMessagingAccountSystem(address);
            this.addAccountSystem(this.messagingSystem);
        }
    }
    /**
     * Create an instance of radix account from an address
     * @param address string address
     * @param [plain] If set to false, will not create default account systems.
     * Use this for accounts that will not be connected to the network
     * @returns
     */
    static fromAddress(address, plain = false) {
        return new RadixAccount(RadixAddress.fromAddress(address), plain);
    }
    /**
     * Create an instance of a Radix account from an arbitrary byte buffer. This
     * could e.g. be a friendly name of an account, in which case it would be
     * created as <code>Buffer.from('friendly name')</code>.
     *
     * @param seed Buffer seed for the address
     * @param [plain] If set to true, will not create default account systems.
     * Use this for accounts that will not be connected to the network.
     * @returns a new Radix account.
     */
    static fromSeed(seed, plain = false) {
        const hash = radixHash(seed);
        return new RadixAccount(RadixAddress.fromPrivate(hash), plain);
    }
    enableDecryption(decryptionProvider) {
        this.decryptionSystem.decryptionProvider = decryptionProvider;
    }
    enableCache(cacheProvider) {
        this.cacheSystem.atomCache = cacheProvider;
        // Load atoms from cache
        return this.cacheSystem.loadAtoms().then((atoms) => {
            for (const atom of atoms) {
                this._onAtomReceived({
                    action: 'STORE',
                    atom,
                    processedData: {},
                    isHead: false,
                });
            }
        });
    }
    getAddress() {
        return this.address.getAddress();
    }
    addAccountSystem(system) {
        if (this.accountSystems.has(system.name)) {
            throw new Error(`System "${system.name}" already exists in account, you can only have one of each system per account`);
        }
        this.accountSystems.set(system.name, system);
        return system;
    }
    removeAccountSystem(name) {
        if (this.accountSystems.has(name)) {
            this.accountSystems.delete(name);
        }
    }
    getSystem(name) {
        if (this.accountSystems.has(name)) {
            return this.accountSystems.get(name);
        }
        throw new Error(`System "${name}" doesn't exist in account`);
    }
    /**
     * An observable that tells you when the account is in sync with the network
     *
     * @returns An observable which sends 'true' whenever the account has received and processed new information form the network
     */
    isSynced() {
        return this.synced.share();
    }
}
