// radixdlt/build/module/modules/account/RadixDataAccountSystem.js
import { Subject, Observable } from 'rxjs';
import { TSMap } from 'typescript-map';
import { filter } from 'rxjs/operators';
import { RadixAddress } from '../atommodel';
export default class RadixDataAccountSystem {
    constructor(address) {
        this.address = address;
        this.name = 'DATA';
        this.applicationDataSubject = new Subject();
        this.applicationData = new TSMap();
    }
    async processAtomUpdate(atomUpdate) {
        if (!('decryptedData' in atomUpdate.processedData)) {
            return;
        }
        if (atomUpdate.action === 'STORE') {
            this.processStoreAtom(atomUpdate);
        }
        else if (atomUpdate.action === 'DELETE') {
            this.processDeleteAtom(atomUpdate);
        }
    }
    processStoreAtom(atomUpdate) {
        const atom = atomUpdate.atom;
        const aid = atom.getAidString();
        const signatures = atom.signatures;
        const applicationId = atomUpdate.processedData.decryptedData.application;
        // Skip existing atoms
        if (this.applicationData.has(applicationId) &&
            this.applicationData.get(applicationId).has(aid)) {
            return;
        }
        const applicationData = {
            aid,
            payload: atomUpdate.processedData.decryptedData,
            timestamp: atom.getTimestamp(),
            signatures,
        };
        const applicationDataUpdate = {
            action: 'STORE',
            aid,
            applicationId,
            data: applicationData,
            signatures,
        };
        if (!this.applicationData.has(applicationId)) {
            this.applicationData.set(applicationId, new TSMap());
        }
        this.applicationData.get(applicationId).set(aid, applicationData);
        this.applicationDataSubject.next(applicationDataUpdate);
    }
    processDeleteAtom(atomUpdate) {
        const atom = atomUpdate.atom;
        const aid = atom.getAidString();
        const signatures = atom.signatures;
        const applicationId = atomUpdate.processedData.decryptedData.application;
        // Skip nonexisting atoms
        if (!this.applicationData.has(applicationId) ||
            !this.applicationData.get(applicationId).has(aid)) {
            return;
        }
        const applicationData = this.applicationData.get(applicationId).get(aid);
        const applicationDataUpdate = {
            action: 'DELETE',
            aid,
            applicationId,
            data: applicationData,
            signatures,
        };
        this.applicationData.get(applicationId).delete(aid);
        this.applicationDataSubject.next(applicationDataUpdate);
    }
    /**
     * Gets application data messages by application id and optionally by signer
     *
     * @param applicationId - Application id of the payload atoms
     * @param [addresses] - List of account addreses to filter application messages by signer
     * @returns An observable subscribed to old and new application messages that met the filter requirements
     */
    getApplicationData(applicationId, addresses) {
        // Pre-calculate signatureIds
        const signatureIds = !addresses ? undefined : addresses.map(a => RadixAddress.fromAddress(a).getUID().toString());
        return Observable.create((observer) => {
            // Send all old data
            if (this.applicationData.has(applicationId)) {
                for (const applicationData of this.applicationData
                    .get(applicationId)
                    .values()) {
                    if (!signatureIds
                        || signatureIds.length === 0
                        || signatureIds.some(s => Object.keys(applicationData.signatures).includes(s))) {
                        const applicationDataUpdate = {
                            action: 'STORE',
                            aid: applicationData.aid,
                            applicationId,
                            data: applicationData,
                            signatures: applicationData.signatures,
                        };
                        observer.next(applicationDataUpdate);
                    }
                }
            }
            // Subscribe for new ones
            this.applicationDataSubject
                .pipe(filter(update => {
                return update.applicationId === applicationId
                    && (!signatureIds
                        || signatureIds.length === 0
                        || signatureIds.some(s => Object.keys(update.signatures).includes(s)));
            }))
                .subscribe(observer);
        });
    }
}
