// radixdlt/build/module/modules/atommodel/particles/tokens/RadixTokenDefinitionParticle.js
import * as tslib_1 from "tslib";
import BN from 'bn.js';
import { includeDSON, includeJSON, RadixParticle, RadixSerializer, RadixFungibleType, RadixAddress, RadixUInt256, RRI, } from '../..';
export var RadixTokenPermissionsValues;
(function (RadixTokenPermissionsValues) {
    RadixTokenPermissionsValues["TOKEN_CREATION_ONLY"] = "token_creation_only";
    RadixTokenPermissionsValues["TOKEN_OWNER_ONLY"] = "token_owner_only";
    RadixTokenPermissionsValues["ALL"] = "all";
    RadixTokenPermissionsValues["NONE"] = "none";
})(RadixTokenPermissionsValues || (RadixTokenPermissionsValues = {}));
/**
 * Particle defining a token
 */
let RadixTokenDefinitionParticle = class RadixTokenDefinitionParticle extends RadixParticle {
    constructor(address, name, symbol, description, granularity, iconUrl, permissions) {
        super();
        this.address = address;
        this.name = name;
        this.symbol = symbol;
        this.description = description;
        this.granularity = new RadixUInt256(granularity);
        this.iconUrl = iconUrl;
        this.permissions = permissions;
    }
    getAddresses() {
        return [this.address];
    }
    getPermissions(action) {
        // Hack because it's 'mint' in permissions but 'minted' in OwnedTokensParticle
        return this.permissions[RadixFungibleType[action].toLowerCase()];
    }
    getGranularity() {
        return this.granularity.value;
    }
    getOwner() {
        return this.address;
    }
    getRRI() {
        return new RRI(this.address, this.symbol);
    }
    getTokenDefinitionReference() {
        return this.getRRI();
    }
    getIconUrl() {
        return this.iconUrl;
    }
};
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", String)
], RadixTokenDefinitionParticle.prototype, "name", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", String)
], RadixTokenDefinitionParticle.prototype, "symbol", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", String)
], RadixTokenDefinitionParticle.prototype, "description", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RadixUInt256)
], RadixTokenDefinitionParticle.prototype, "granularity", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", Object)
], RadixTokenDefinitionParticle.prototype, "permissions", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RadixAddress)
], RadixTokenDefinitionParticle.prototype, "address", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", String)
], RadixTokenDefinitionParticle.prototype, "iconUrl", void 0);
RadixTokenDefinitionParticle = tslib_1.__decorate([
    RadixSerializer.registerClass('radix.particles.token_definition'),
    tslib_1.__metadata("design:paramtypes", [RadixAddress, String, String, String, BN, String, Object])
], RadixTokenDefinitionParticle);
export { RadixTokenDefinitionParticle };
