// radixdlt/build/module/modules/atommodel/particles/tokens/RadixTransferrableTokensParticle.js
import * as tslib_1 from "tslib";
import { includeDSON, includeJSON, RadixSerializer, RadixParticle, RadixAddress, RadixFungibleType, RadixUInt256, RRI, } from '../..';
import BN from 'bn.js';
/**
 *  A particle which represents an amount of consumable and consuming, tranferable fungible tokens
 *  owned by some key owner and stored in an account.
 */
let RadixTransferrableTokensParticle = class RadixTransferrableTokensParticle extends RadixParticle {
    constructor(amount, granularity, address, nonce, tokenReference, tokenPermissions, planck) {
        if (amount.isZero()) {
            throw new Error('Ammount cannot be zero');
        }
        planck = planck ? planck : Math.floor(Date.now() / 60000 + 60000);
        super();
        this.address = address;
        this.granularity = new RadixUInt256(granularity);
        this.tokenDefinitionReference = tokenReference;
        this.amount = new RadixUInt256(amount);
        this.planck = planck;
        this.nonce = nonce;
        this.permissions = tokenPermissions;
    }
    getAddress() {
        return this.address;
    }
    getAddresses() {
        return [this.address];
    }
    getType() {
        return RadixFungibleType.TRANSFER;
    }
    getPlanck() {
        return this.planck;
    }
    getNonce() {
        return this.nonce;
    }
    getTokenDefinitionReference() {
        return this.tokenDefinitionReference;
    }
    getOwner() {
        return this.address;
    }
    getAmount() {
        return this.amount.value;
    }
    getGranularity() {
        return this.granularity.value;
    }
    getTokenPermissions() {
        return this.permissions;
    }
};
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RadixAddress)
], RadixTransferrableTokensParticle.prototype, "address", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RRI)
], RadixTransferrableTokensParticle.prototype, "tokenDefinitionReference", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RadixUInt256)
], RadixTransferrableTokensParticle.prototype, "granularity", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", Number)
], RadixTransferrableTokensParticle.prototype, "planck", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", Number)
], RadixTransferrableTokensParticle.prototype, "nonce", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RadixUInt256)
], RadixTransferrableTokensParticle.prototype, "amount", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", Object)
], RadixTransferrableTokensParticle.prototype, "permissions", void 0);
RadixTransferrableTokensParticle = tslib_1.__decorate([
    RadixSerializer.registerClass('radix.particles.transferrable_tokens'),
    tslib_1.__metadata("design:paramtypes", [BN,
        BN,
        RadixAddress, Number, RRI, Object, Number])
], RadixTransferrableTokensParticle);
export { RadixTransferrableTokensParticle };
