// radixdlt/build/module/modules/atommodel/particles/tokens/RadixUnallocatedTokensParticle.js
import * as tslib_1 from "tslib";
import { includeDSON, includeJSON, RadixSerializer, RadixParticle, RadixFungibleType, RadixUInt256, RRI, } from '../..';
import BN from 'bn.js';
/**
 *  A particle which represents an amount of unallocated tokens which can be minted.
 */
let RadixUnallocatedTokensParticle = class RadixUnallocatedTokensParticle extends RadixParticle {
    constructor(amount, granularity, nonce, tokenReference, tokenPermissions) {
        if (amount.isZero()) {
            throw new Error('Ammount cannot be zero');
        }
        super();
        this.granularity = new RadixUInt256(granularity);
        this.tokenDefinitionReference = tokenReference;
        this.amount = new RadixUInt256(amount);
        this.nonce = nonce;
        this.permissions = tokenPermissions;
    }
    getAddress() {
        return this.tokenDefinitionReference.address;
    }
    getAddresses() {
        return [this.tokenDefinitionReference.address];
    }
    getType() {
        return RadixFungibleType.TRANSFER;
    }
    getNonce() {
        return this.nonce;
    }
    getTokenDefinitionReference() {
        return this.tokenDefinitionReference;
    }
    getOwner() {
        return this.getAddress();
    }
    getAmount() {
        return this.amount.value;
    }
    getGranularity() {
        return this.granularity.value;
    }
    getTokenPermissions() {
        return this.permissions;
    }
};
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RRI)
], RadixUnallocatedTokensParticle.prototype, "tokenDefinitionReference", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RadixUInt256)
], RadixUnallocatedTokensParticle.prototype, "granularity", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", Number)
], RadixUnallocatedTokensParticle.prototype, "nonce", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RadixUInt256)
], RadixUnallocatedTokensParticle.prototype, "amount", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", Object)
], RadixUnallocatedTokensParticle.prototype, "permissions", void 0);
RadixUnallocatedTokensParticle = tslib_1.__decorate([
    RadixSerializer.registerClass('radix.particles.unallocated_tokens'),
    tslib_1.__metadata("design:paramtypes", [BN,
        BN, Number, RRI, Object])
], RadixUnallocatedTokensParticle);
export { RadixUnallocatedTokensParticle };
