// radixdlt/build/module/modules/messaging/RadixMessagingAccountSystem.js
import { Subject, Observable } from 'rxjs';
import { TSMap } from 'typescript-map';
import { RadixDecryptionState } from '../account/RadixDecryptionAccountSystem';
export default class RadixMessagingAccountSystem {
    constructor(address) {
        this.address = address;
        this.name = 'RADIX-MESSAGING';
        this.messageSubject = new Subject();
        this.chats = new TSMap();
        this.messages = new TSMap();
    }
    async processAtomUpdate(atomUpdate) {
        if (!('decryptedData' in atomUpdate.processedData) ||
            atomUpdate.processedData.decryptedData.application !== 'message' ||
            atomUpdate.processedData.decryptedData.decryptionState === RadixDecryptionState.CANNOT_DECRYPT) {
            return;
        }
        if (atomUpdate.action === 'STORE') {
            this.processStoreAtom(atomUpdate);
        }
        else if (atomUpdate.action === 'DELETE') {
            this.processDeleteAtom(atomUpdate);
        }
    }
    startNewChat(to) {
        // Create new chat
        const chatId = to.getAddress();
        if (this.chats.has(chatId)) {
            return;
        }
        const chatDescription = {
            address: to.getAddress(),
            chat_id: chatId,
            title: chatId,
            last_message_timestamp: Date.now(),
            messages: new TSMap(),
        };
        // Add at the top
        this.chats.set(chatId, chatDescription);
    }
    processStoreAtom(atomUpdate) {
        const atom = atomUpdate.atom;
        const aid = atom.getAidString();
        const signatures = atom.signatures;
        const decryptedData = atomUpdate.processedData.decryptedData;
        // Skip existing atoms
        if (this.messages.has(aid)) {
            return;
        }
        const from = decryptedData.from;
        const to = decryptedData.to.find(a => !a.equals(from));
        if (!to) {
            throw new Error('A message needs to have at least one other recipient');
        }
        // Chat id
        let address = null;
        let isMine = false;
        if (from.equals(this.address)) {
            address = to;
            isMine = true;
        }
        else {
            address = from;
        }
        if (address === null) {
            throw new Error('Error processing a radix-message atom: neither of addresses is owned by this account');
        }
        const chatId = address.toString();
        const message = {
            aid,
            chat_id: chatId,
            to,
            from,
            content: decryptedData.data,
            timestamp: atom.getTimestamp(),
            is_mine: isMine,
            encryptionState: decryptedData.decryptionState,
        };
        // Find existing chat
        // Otherwise create new chat
        if (!this.chats.has(chatId)) {
            const newChatDescription = {
                address: address.getAddress(),
                chat_id: chatId,
                title: chatId,
                last_message_timestamp: message.timestamp,
                messages: new TSMap(),
            };
            this.chats.set(chatId, newChatDescription);
        }
        const chatDescription = this.chats.get(chatId);
        if (message.timestamp > chatDescription.last_message_timestamp) {
            chatDescription.last_message_timestamp = message.timestamp;
        }
        chatDescription.messages.set(aid, message);
        // Move chat to the top
        this.chats.delete(chatId);
        this.chats.set(chatId, chatDescription);
        this.messages.set(aid, message);
        const messageUpdate = {
            action: 'STORE',
            aid,
            message,
        };
        this.messageSubject.next(messageUpdate);
    }
    processDeleteAtom(atomUpdate) {
        const atom = atomUpdate.atom;
        const aid = atom.getAidString();
        // Skip nonexisting atoms
        if (!this.messages.has(aid)) {
            return;
        }
        const message = this.messages.get(aid);
        this.chats.get(message.chat_id).messages.delete(aid);
        this.messages.delete(aid);
        const messageUpdate = {
            action: 'DELETE',
            aid,
            message,
        };
        this.messageSubject.next(messageUpdate);
    }
    getAllMessages() {
        return Observable.create((observer) => {
            // Send all old transactions
            for (const message of this.messages.values()) {
                const messageUpdate = {
                    action: 'STORE',
                    aid: message.aid,
                    message,
                };
                observer.next(messageUpdate);
            }
            // Subscribe for new ones
            this.messageSubject.subscribe(observer);
        });
    }
}
