// radixdlt/build/main/modules/atommodel/particles/tokens/RadixUnallocatedTokensParticle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("../..");
var bn_js_1 = tslib_1.__importDefault(require("bn.js"));
/**
 *  A particle which represents an amount of unallocated tokens which can be minted.
 */
var RadixUnallocatedTokensParticle = /** @class */ (function (_super) {
    tslib_1.__extends(RadixUnallocatedTokensParticle, _super);
    function RadixUnallocatedTokensParticle(amount, granularity, nonce, tokenReference, tokenPermissions) {
        var _this = this;
        if (amount.isZero()) {
            throw new Error('Ammount cannot be zero');
        }
        _this = _super.call(this) || this;
        _this.granularity = new __1.RadixUInt256(granularity);
        _this.tokenDefinitionReference = tokenReference;
        _this.amount = new __1.RadixUInt256(amount);
        _this.nonce = nonce;
        _this.permissions = tokenPermissions;
        return _this;
    }
    RadixUnallocatedTokensParticle.prototype.getAddress = function () {
        return this.tokenDefinitionReference.address;
    };
    RadixUnallocatedTokensParticle.prototype.getAddresses = function () {
        return [this.tokenDefinitionReference.address];
    };
    RadixUnallocatedTokensParticle.prototype.getType = function () {
        return __1.RadixFungibleType.TRANSFER;
    };
    RadixUnallocatedTokensParticle.prototype.getNonce = function () {
        return this.nonce;
    };
    RadixUnallocatedTokensParticle.prototype.getTokenDefinitionReference = function () {
        return this.tokenDefinitionReference;
    };
    RadixUnallocatedTokensParticle.prototype.getOwner = function () {
        return this.getAddress();
    };
    RadixUnallocatedTokensParticle.prototype.getAmount = function () {
        return this.amount.value;
    };
    RadixUnallocatedTokensParticle.prototype.getGranularity = function () {
        return this.granularity.value;
    };
    RadixUnallocatedTokensParticle.prototype.getTokenPermissions = function () {
        return this.permissions;
    };
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RRI)
    ], RadixUnallocatedTokensParticle.prototype, "tokenDefinitionReference", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RadixUInt256)
    ], RadixUnallocatedTokensParticle.prototype, "granularity", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", Number)
    ], RadixUnallocatedTokensParticle.prototype, "nonce", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RadixUInt256)
    ], RadixUnallocatedTokensParticle.prototype, "amount", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", Object)
    ], RadixUnallocatedTokensParticle.prototype, "permissions", void 0);
    RadixUnallocatedTokensParticle = tslib_1.__decorate([
        __1.RadixSerializer.registerClass('radix.particles.unallocated_tokens'),
        tslib_1.__metadata("design:paramtypes", [bn_js_1.default,
            bn_js_1.default, Number, __1.RRI, Object])
    ], RadixUnallocatedTokensParticle);
    return RadixUnallocatedTokensParticle;
}(__1.RadixParticle));
exports.RadixUnallocatedTokensParticle = RadixUnallocatedTokensParticle;
