// radixdlt/build/main/modules/atommodel/particles/tokens/RadixTransferrableTokensParticle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("../..");
var bn_js_1 = tslib_1.__importDefault(require("bn.js"));
/**
 *  A particle which represents an amount of consumable and consuming, tranferable fungible tokens
 *  owned by some key owner and stored in an account.
 */
var RadixTransferrableTokensParticle = /** @class */ (function (_super) {
    tslib_1.__extends(RadixTransferrableTokensParticle, _super);
    function RadixTransferrableTokensParticle(amount, granularity, address, nonce, tokenReference, tokenPermissions, planck) {
        var _this = this;
        if (amount.isZero()) {
            throw new Error('Ammount cannot be zero');
        }
        planck = planck ? planck : Math.floor(Date.now() / 60000 + 60000);
        _this = _super.call(this) || this;
        _this.address = address;
        _this.granularity = new __1.RadixUInt256(granularity);
        _this.tokenDefinitionReference = tokenReference;
        _this.amount = new __1.RadixUInt256(amount);
        _this.planck = planck;
        _this.nonce = nonce;
        _this.permissions = tokenPermissions;
        return _this;
    }
    RadixTransferrableTokensParticle.prototype.getAddress = function () {
        return this.address;
    };
    RadixTransferrableTokensParticle.prototype.getAddresses = function () {
        return [this.address];
    };
    RadixTransferrableTokensParticle.prototype.getType = function () {
        return __1.RadixFungibleType.TRANSFER;
    };
    RadixTransferrableTokensParticle.prototype.getPlanck = function () {
        return this.planck;
    };
    RadixTransferrableTokensParticle.prototype.getNonce = function () {
        return this.nonce;
    };
    RadixTransferrableTokensParticle.prototype.getTokenDefinitionReference = function () {
        return this.tokenDefinitionReference;
    };
    RadixTransferrableTokensParticle.prototype.getOwner = function () {
        return this.address;
    };
    RadixTransferrableTokensParticle.prototype.getAmount = function () {
        return this.amount.value;
    };
    RadixTransferrableTokensParticle.prototype.getGranularity = function () {
        return this.granularity.value;
    };
    RadixTransferrableTokensParticle.prototype.getTokenPermissions = function () {
        return this.permissions;
    };
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RadixAddress)
    ], RadixTransferrableTokensParticle.prototype, "address", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RRI)
    ], RadixTransferrableTokensParticle.prototype, "tokenDefinitionReference", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RadixUInt256)
    ], RadixTransferrableTokensParticle.prototype, "granularity", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", Number)
    ], RadixTransferrableTokensParticle.prototype, "planck", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", Number)
    ], RadixTransferrableTokensParticle.prototype, "nonce", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RadixUInt256)
    ], RadixTransferrableTokensParticle.prototype, "amount", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", Object)
    ], RadixTransferrableTokensParticle.prototype, "permissions", void 0);
    RadixTransferrableTokensParticle = tslib_1.__decorate([
        __1.RadixSerializer.registerClass('radix.particles.transferrable_tokens'),
        tslib_1.__metadata("design:paramtypes", [bn_js_1.default,
            bn_js_1.default,
            __1.RadixAddress, Number, __1.RRI, Object, Number])
    ], RadixTransferrableTokensParticle);
    return RadixTransferrableTokensParticle;
}(__1.RadixParticle));
exports.RadixTransferrableTokensParticle = RadixTransferrableTokensParticle;
