// radixdlt/build/main/modules/atommodel/particles/tokens/RadixTokenDefinitionParticle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var bn_js_1 = tslib_1.__importDefault(require("bn.js"));
var __1 = require("../..");
var RadixTokenPermissionsValues;
(function (RadixTokenPermissionsValues) {
    RadixTokenPermissionsValues["TOKEN_CREATION_ONLY"] = "token_creation_only";
    RadixTokenPermissionsValues["TOKEN_OWNER_ONLY"] = "token_owner_only";
    RadixTokenPermissionsValues["ALL"] = "all";
    RadixTokenPermissionsValues["NONE"] = "none";
})(RadixTokenPermissionsValues = exports.RadixTokenPermissionsValues || (exports.RadixTokenPermissionsValues = {}));
/**
 * Particle defining a token
 */
var RadixTokenDefinitionParticle = /** @class */ (function (_super) {
    tslib_1.__extends(RadixTokenDefinitionParticle, _super);
    function RadixTokenDefinitionParticle(address, name, symbol, description, granularity, iconUrl, permissions) {
        var _this = _super.call(this) || this;
        _this.address = address;
        _this.name = name;
        _this.symbol = symbol;
        _this.description = description;
        _this.granularity = new __1.RadixUInt256(granularity);
        _this.iconUrl = iconUrl;
        _this.permissions = permissions;
        return _this;
    }
    RadixTokenDefinitionParticle.prototype.getAddresses = function () {
        return [this.address];
    };
    RadixTokenDefinitionParticle.prototype.getPermissions = function (action) {
        // Hack because it's 'mint' in permissions but 'minted' in OwnedTokensParticle
        return this.permissions[__1.RadixFungibleType[action].toLowerCase()];
    };
    RadixTokenDefinitionParticle.prototype.getGranularity = function () {
        return this.granularity.value;
    };
    RadixTokenDefinitionParticle.prototype.getOwner = function () {
        return this.address;
    };
    RadixTokenDefinitionParticle.prototype.getRRI = function () {
        return new __1.RRI(this.address, this.symbol);
    };
    RadixTokenDefinitionParticle.prototype.getTokenDefinitionReference = function () {
        return this.getRRI();
    };
    RadixTokenDefinitionParticle.prototype.getIconUrl = function () {
        return this.iconUrl;
    };
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", String)
    ], RadixTokenDefinitionParticle.prototype, "name", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", String)
    ], RadixTokenDefinitionParticle.prototype, "symbol", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", String)
    ], RadixTokenDefinitionParticle.prototype, "description", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RadixUInt256)
    ], RadixTokenDefinitionParticle.prototype, "granularity", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", Object)
    ], RadixTokenDefinitionParticle.prototype, "permissions", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", __1.RadixAddress)
    ], RadixTokenDefinitionParticle.prototype, "address", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", String)
    ], RadixTokenDefinitionParticle.prototype, "iconUrl", void 0);
    RadixTokenDefinitionParticle = tslib_1.__decorate([
        __1.RadixSerializer.registerClass('radix.particles.token_definition'),
        tslib_1.__metadata("design:paramtypes", [__1.RadixAddress, String, String, String, bn_js_1.default, String, Object])
    ], RadixTokenDefinitionParticle);
    return RadixTokenDefinitionParticle;
}(__1.RadixParticle));
exports.RadixTokenDefinitionParticle = RadixTokenDefinitionParticle;
