// radixdlt/build/module/modules/atommodel/particles/RadixUniqueParticle.js
import * as tslib_1 from "tslib";
import { RadixSerializer, RadixParticle, includeDSON, includeJSON, RadixAddress, RRI } from '..';
let RadixUniqueParticle = class RadixUniqueParticle extends RadixParticle {
    constructor(address, unique, nonce) {
        super();
        this.address = address;
        this.name = unique;
        this.nonce = nonce ? nonce : Date.now();
    }
    getAddresses() {
        return [this.address];
    }
    getRRI() {
        return new RRI(this.address, this.name);
    }
};
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", String)
], RadixUniqueParticle.prototype, "name", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", RadixAddress)
], RadixUniqueParticle.prototype, "address", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", Number)
], RadixUniqueParticle.prototype, "nonce", void 0);
RadixUniqueParticle = tslib_1.__decorate([
    RadixSerializer.registerClass('radix.particles.unique'),
    tslib_1.__metadata("design:paramtypes", [RadixAddress, String, Number])
], RadixUniqueParticle);
export { RadixUniqueParticle };
