// radixdlt/build/module/modules/atommodel/particles/RadixParticle.js
import * as tslib_1 from "tslib";
import { RadixSerializableObject, RadixSerializer, includeJSON, includeDSON, RadixEUID } from '..';
import { StringifySet } from '../../common/StringifySet';
let RadixParticle = class RadixParticle extends RadixSerializableObject {
    constructor() {
        super();
    }
    getAddresses() {
        throw new Error('Particle implementations must override this method!');
    }
    getHid() {
        return new RadixEUID(this.getHash().slice(0, 16));
    }
    getHidString() {
        return this.getHid().toString();
    }
    getDestinations() {
        return StringifySet.of(this.getAddresses().map(a => a.getUID())).values();
    }
    get destinations() {
        return this.getDestinations();
    }
    set destinations(destinations) {
        // Ignore, since it's a computed property
    }
};
tslib_1.__decorate([
    includeJSON,
    includeDSON,
    tslib_1.__metadata("design:type", Array),
    tslib_1.__metadata("design:paramtypes", [Array])
], RadixParticle.prototype, "destinations", null);
RadixParticle = tslib_1.__decorate([
    RadixSerializer.registerClass('radix.particle'),
    tslib_1.__metadata("design:paramtypes", [])
], RadixParticle);
export { RadixParticle };
