// radixdlt/build/module/modules/atommodel/particles/RadixSpunParticle.js
import * as tslib_1 from "tslib";
import { RadixParticle, RadixSerializableObject, RadixSerializer, includeJSON, includeDSON } from '..';
export var RadixSpin;
(function (RadixSpin) {
    RadixSpin[RadixSpin["UP"] = 1] = "UP";
    RadixSpin[RadixSpin["DOWN"] = -1] = "DOWN";
})(RadixSpin || (RadixSpin = {}));
let RadixSpunParticle = class RadixSpunParticle extends RadixSerializableObject {
    constructor(particle, spin) {
        super();
        this.particle = particle;
        this.spin = spin;
    }
    static up(particle) {
        return new this(particle, RadixSpin.UP);
    }
    static down(particle) {
        return new this(particle, RadixSpin.DOWN);
    }
};
tslib_1.__decorate([
    includeJSON,
    includeDSON,
    tslib_1.__metadata("design:type", RadixParticle)
], RadixSpunParticle.prototype, "particle", void 0);
tslib_1.__decorate([
    includeJSON,
    includeDSON,
    tslib_1.__metadata("design:type", Number)
], RadixSpunParticle.prototype, "spin", void 0);
RadixSpunParticle = tslib_1.__decorate([
    RadixSerializer.registerClass('radix.spun_particle'),
    tslib_1.__metadata("design:paramtypes", [RadixParticle, Number])
], RadixSpunParticle);
export { RadixSpunParticle };
