// radixdlt/build/module/modules/atommodel/serializer/RadixSerializer.js
import cbor from 'cbor';
import { TSMap } from 'typescript-map';
import { logger } from '../../common/RadixLogger';
import 'reflect-metadata';
import { isEmpty } from '../../common/RadixUtil';
export const JSON_PROPERTIES_KEY = 'JSON_SERIALIZATION_PROPERTIES';
export const DSON_PROPERTIES_KEY = 'DSON_SERIALIZATION_PROPERTIES';
/**
 * Decorator to register a property for JSON serialization.
 *   Stores property names in the prototype metadata
 * @param target
 * @param propertyName
 */
export function includeJSON(target, propertyName) {
    registerPropertyForSerialization(target, propertyName, JSON_PROPERTIES_KEY);
}
/**
 * Decorator to register a property for DSON serialization.
 *   Stores property names in the prototype metadata
 * @param target
 * @param propertyName
 */
export function includeDSON(target, propertyName) {
    registerPropertyForSerialization(target, propertyName, DSON_PROPERTIES_KEY);
}
/**
 * Registers property for serialization.
 *   Stores an array of property names in the prototype metadata.
 *   Inherits from the object prototype chain.
 * @param target
 * @param propertyName
 * @param metadataKey
 */
function registerPropertyForSerialization(target, propertyName, metadataKey) {
    // Avoid modifying parent metadata, clone instead
    if (!Reflect.hasOwnMetadata(metadataKey, target)) {
        let props = [];
        if (Reflect.hasMetadata(metadataKey, target)) {
            // Clone parent metadata
            props = Reflect.getMetadata(metadataKey, target).slice(0);
        }
        Reflect.defineMetadata(metadataKey, props, target);
    }
    Reflect.getMetadata(metadataKey, target).push(propertyName);
    Reflect.getMetadata(metadataKey, target).sort();
}
export class RadixSerializer {
    /**
     * Decorator to register a class for serialization
     * @param serializer serializer id
     * @returns
     */
    static registerClass(serializer) {
        return (constructor) => {
            constructor.SERIALIZER = serializer;
            this.classes.set(serializer, constructor);
        };
    }
    /**
     * Registers advanced 'primitive' types for serialization
     * @param tag
     * @returns
     */
    static registerPrimitive(tag) {
        return (constructor) => {
            this.primitives.set(tag, constructor);
        };
    }
    static toJSON(data) {
        if (Array.isArray(data)) {
            const output = [];
            for (const item of data) {
                output.push(RadixSerializer.toJSON(item));
            }
            return output;
        }
        else if (typeof data === 'number' ||
            typeof data === 'boolean') {
            return data;
        }
        else if (typeof data === 'string') {
            return `:str:${data}`;
        }
        else if (data !== null && typeof data === 'object') {
            if (typeof data.toJSON === 'function') {
                return data.toJSON();
            }
            else {
                const output = {};
                for (const key in data) {
                    const serialized = RadixSerializer.toJSON(data[key]);
                    if (!isEmpty(serialized)) {
                        output[key] = serialized;
                    }
                }
                return output;
            }
        }
    }
    static fromJSON(json) {
        if (Array.isArray(json)) {
            const output = [];
            for (const item of json) {
                output.push(RadixSerializer.fromJSON(item));
            }
            return output;
        }
        else if (typeof json === 'object') {
            return RadixSerializer.fromJSONObject(json);
        }
        else if (typeof json === 'string') {
            // Advanced primitives
            const tag = json.slice(0, 5);
            if (tag === ':str:') {
                return json.slice(5);
            }
            // Cast to a primitve class
            if (this.primitives.has(tag)) {
                return this.primitives.get(tag).fromJSON(json.slice(5));
            }
            logger.warn(`No matching class for primitive string "${json}"`);
        }
        else {
            return json;
        }
    }
    static fromJSONObject(jsonObject) {
        const output = {};
        for (const key in jsonObject) {
            output[key] = (key === 'serializer' ? jsonObject[key] : RadixSerializer.fromJSON(jsonObject[key]));
        }
        if ('serializer' in output) {
            // tslint:disable-next-line:no-string-literal
            const type = output['serializer'];
            if (this.classes.has(type)) {
                return this.classes.get(type).fromJSON(output);
            }
            logger.warn(`Serializer "${type}" not implemented`);
        }
        return output;
    }
    static toDSON(data) {
        const enc = new cbor.Encoder();
        // Overide default object encoder to use stream encoding and lexicographical ordering of keys
        enc.addSemanticType(Object, (encoder, obj) => {
            const keys = Object.keys(obj);
            keys.sort();
            if (!encoder.push(Buffer.from([191]))) {
                return false;
            }
            for (const key of keys) {
                if (isEmpty(obj[key])) {
                    continue;
                }
                if (!encoder.pushAny(key)) {
                    return false;
                }
                if (!encoder.pushAny(obj[key])) {
                    return false;
                }
            }
            if (!encoder.push(Buffer.from([0xFF]))) {
                return false;
            }
            return true;
        });
        return enc._encodeAll([data]);
    }
}
RadixSerializer.classes = new TSMap();
RadixSerializer.primitives = new TSMap();
