// radixdlt/build/main/modules/common/RadixUtil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var long_1 = tslib_1.__importDefault(require("long"));
var bn_js_1 = tslib_1.__importDefault(require("bn.js"));
var crypto_1 = tslib_1.__importDefault(require("crypto"));
function radixHash(data, offset, len) {
    if (offset) {
        data = data.slice(offset, len);
    }
    if (!Buffer.isBuffer(data)) {
        data = Buffer.from(data);
    }
    // Double hash to protect against length extension attacks
    var hash1 = crypto_1.default.createHash('sha256');
    hash1.update(data);
    var hash2 = crypto_1.default.createHash('sha256');
    hash2.update(hash1.digest());
    return hash2.digest();
}
exports.radixHash = radixHash;
function bigIntFromByteArray(bytes) {
    return new bn_js_1.default(bytes).fromTwos(bytes.length * 8);
}
exports.bigIntFromByteArray = bigIntFromByteArray;
function byteArrayFromBigInt(num) {
    // Compatibility with Java BigInteger.toByteArray() https://stackoverflow.com/a/24158695
    var byteLength = Math.ceil((num.bitLength() + 1) / 8);
    var result = num.toTwos(8 * byteLength).toArrayLike(Buffer);
    if (result.length !== byteLength) {
        var newResult = Buffer.alloc(byteLength, 0);
        result.copy(newResult, byteLength - result.length);
        return newResult;
    }
    return result;
}
exports.byteArrayFromBigInt = byteArrayFromBigInt;
function longFromBigInt(num) {
    // Emulate Java BigInteger.longValue(), following the spec at 5.1.3 https://docs.oracle.com/javase/specs/jls/se7/html/jls-5.html
    var byteLength = Math.max(8, num.byteLength());
    var bytes = num.toTwos(8 * byteLength).toArray('be', byteLength);
    var truncatedBytes = bytes.slice(bytes.length - 8, bytes.length);
    return long_1.default.fromBytesBE(truncatedBytes);
}
exports.longFromBigInt = longFromBigInt;
function bigIntFromLong(num) {
    return new bn_js_1.default(num.toBytesBE(), 'be').fromTwos(64);
}
exports.bigIntFromLong = bigIntFromLong;
function powTargetFromAtomSize(size) {
    var target = Buffer.alloc(32, 0xff);
    var leadingBits = Math.ceil(Math.log(size * 8));
    var leadingBytes = Math.floor(leadingBits / 8);
    var leftOverBits = leadingBits % 8;
    target.fill(0, 0, leadingBytes);
    var middleByte = ~(0xff << (8 - leftOverBits)) & 0xff;
    target.writeUInt8(middleByte, leadingBytes);
    return target;
}
exports.powTargetFromAtomSize = powTargetFromAtomSize;
function shuffleArray(arr) {
    return arr
        .map(function (a) { return [Math.random(), a]; })
        .sort(function (a, b) { return a[0] - b[0]; })
        .map(function (a) { return a[1]; });
}
exports.shuffleArray = shuffleArray;
function isEmpty(val) {
    return val === undefined
        || val === null
        || val.length === 0
        || (Object.keys(val).length === 0 && val.constructor === Object);
}
exports.isEmpty = isEmpty;
