// radixdlt/build/main/modules/pow/RadixPOWTask.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var RadixPOW_1 = tslib_1.__importDefault(require("./RadixPOW"));
var RadixPOWTask = /** @class */ (function () {
    function RadixPOWTask(magic, seed, target) {
        this.magic = magic;
        this.seed = seed;
        this.target = target;
        this.pow = new RadixPOW_1.default(magic, seed);
    }
    RadixPOWTask.prototype.computePow = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.attemptPow(resolve);
        });
    };
    RadixPOWTask.prototype.attemptPow = function (callback) {
        var _this = this;
        for (var i = 0; i < 1000; i++) {
            this.pow.incrementNonce();
            var hash = this.pow.getHash();
            if (this.meetsTarget(hash)) {
                setTimeout(function () {
                    callback(_this.pow);
                });
                return;
            }
        }
        // Non-blocking
        setTimeout(function () {
            _this.attemptPow(callback);
        }, 0);
    };
    RadixPOWTask.prototype.meetsTarget = function (hash) {
        return hash.compare(this.target) < 0;
    };
    return RadixPOWTask;
}());
exports.default = RadixPOWTask;
