// radixdlt/build/module/modules/atommodel/atom/RadixAtom.js
var RadixAtom_1;
import * as tslib_1 from "tslib";
import { includeJSON, includeDSON, RadixSerializableObject, RadixSerializer, RadixAID, } from '..';
import { StringifySet } from '../../common/StringifySet';
let RadixAtom = RadixAtom_1 = class RadixAtom extends RadixSerializableObject {
    constructor() {
        super(...arguments);
        this.particleGroups = [];
        this.metaData = {};
    }
    getParticles() {
        const particles = [];
        for (const particleGroup of this.particleGroups) {
            for (const particle of particleGroup.particles) {
                particles.push(particle);
            }
        }
        return particles;
    }
    getAddresses() {
        const addressSet = new StringifySet();
        for (const particleGroup of this.particleGroups) {
            for (const particle of particleGroup.particles) {
                const addresses = particle.particle.getAddresses();
                for (const address of addresses) {
                    addressSet.add(address);
                }
            }
        }
        return addressSet.values();
    }
    getShards() {
        const shardSet = new StringifySet();
        for (const particleGroup of this.particleGroups) {
            for (const particle of particleGroup.particles) {
                const addresses = particle.particle.getAddresses();
                for (const address of addresses) {
                    const shard = address.getShard();
                    shardSet.add(shard);
                }
            }
        }
        return shardSet.values();
    }
    getTimestamp() {
        const timestamp = parseInt(this.metaData[RadixAtom_1.METADATA_TIMESTAMP_KEY], 10);
        if (Number.isNaN(timestamp)) {
            throw new Error('Timestamp is not set or not a valid number');
        }
        else {
            return timestamp;
        }
    }
    setTimestamp(timestamp) {
        this.metaData[RadixAtom_1.METADATA_TIMESTAMP_KEY] = '' + timestamp;
    }
    setPowNonce(nonce) {
        this.metaData[RadixAtom_1.METADATA_POW_NONCE_KEY] = nonce.toString();
    }
    clearPowNonce() {
        delete this.metaData[RadixAtom_1.METADATA_POW_NONCE_KEY];
    }
    getSpunParticlesOfType(...types) {
        return this.getParticles()
            .filter(s => {
            for (const type of types) {
                if (s.particle instanceof type) {
                    return true;
                }
            }
        });
    }
    getParticlesOfSpin(spin) {
        return this.getParticles()
            .filter(spunParticle => spunParticle.spin === spin)
            .map(spunParticle => spunParticle.particle);
    }
    getParticlesOfType(type, spin) {
        let particles = this.getParticles()
            .filter(spunParticle => spunParticle.particle instanceof type);
        if (spin) {
            particles = particles.filter(spunParticle => spunParticle.spin === spin);
        }
        return particles.map(spunParticle => spunParticle.particle);
    }
    getFirstParticleOfType(type) {
        return this.getParticles()
            .find(spunParticle => spunParticle.particle instanceof type).particle;
    }
    containsParticle(...types) {
        for (const spunParticle of this.getParticles()) {
            for (const type of types) {
                if (spunParticle.particle instanceof type) {
                    return true;
                }
            }
        }
        return false;
    }
    getParticleGroups() {
        return this.particleGroups;
    }
    getAid() {
        return RadixAID.from(this.getHash(), this.getShards());
    }
    getAidString() {
        return this.getAid().toString();
    }
};
RadixAtom.METADATA_TIMESTAMP_KEY = 'timestamp';
RadixAtom.METADATA_POW_NONCE_KEY = 'powNonce';
tslib_1.__decorate([
    includeJSON,
    includeDSON,
    tslib_1.__metadata("design:type", Array)
], RadixAtom.prototype, "particleGroups", void 0);
tslib_1.__decorate([
    includeJSON,
    tslib_1.__metadata("design:type", Object)
], RadixAtom.prototype, "signatures", void 0);
tslib_1.__decorate([
    includeDSON,
    includeJSON,
    tslib_1.__metadata("design:type", Object)
], RadixAtom.prototype, "metaData", void 0);
RadixAtom = RadixAtom_1 = tslib_1.__decorate([
    RadixSerializer.registerClass('radix.atom')
], RadixAtom);
export { RadixAtom };
