// radixdlt/build/module/modules/atommodel/RadixSerializableObject.js
import * as tslib_1 from "tslib";
import { RadixSerializer, includeJSON, includeDSON, JSON_PROPERTIES_KEY, DSON_PROPERTIES_KEY } from '.';
import { radixHash, isEmpty } from '../common/RadixUtil';
export class RadixSerializableObject {
    constructor(...args) {
        this.version = 100;
        // Empty constructor
    }
    static fromJSON(json) {
        // So that we can have constructors for the different classes
        const obj = Object.create(this.prototype);
        if (json) {
            for (const key in json) {
                if (key === 'constructor' || key === 'serializationProperties') {
                    continue;
                }
                obj[key] = json[key];
            }
        }
        return obj;
    }
    get serializer() {
        return this.constructor.SERIALIZER;
    }
    set serializer(_) {
        // Do nothing
    }
    toJSON() {
        const constructor = this.constructor;
        const output = { serializer: constructor.SERIALIZER };
        const serializationProps = Reflect.getMetadata(JSON_PROPERTIES_KEY, this);
        for (const key of serializationProps) {
            const serialized = RadixSerializer.toJSON(this[key]);
            if (!isEmpty(serialized)) {
                output[key] = serialized;
            }
        }
        return output;
    }
    toDSON() {
        return RadixSerializer.toDSON(this);
    }
    encodeCBOR(encoder) {
        // Streaming encoding for maps
        const serializationProps = Reflect.getMetadata(DSON_PROPERTIES_KEY, this);
        if (!encoder.push(Buffer.from([191]))) {
            return false;
        }
        for (const prop of serializationProps) {
            if (isEmpty(this[prop])) {
                continue;
            }
            if (!encoder.pushAny(prop)) {
                return false;
            }
            if (!encoder.pushAny(this[prop])) {
                return false;
            }
        }
        if (!encoder.push(Buffer.from([0xFF]))) {
            return false;
        }
        return true;
    }
    getHash() {
        return radixHash(this.toDSON());
    }
    getSize() {
        return this.toDSON().length;
    }
}
RadixSerializableObject.SERIALIZER = '';
tslib_1.__decorate([
    includeJSON,
    includeDSON,
    tslib_1.__metadata("design:type", Object)
], RadixSerializableObject.prototype, "version", void 0);
tslib_1.__decorate([
    includeDSON,
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], RadixSerializableObject.prototype, "serializer", null);
