// radixdlt/build/module/modules/token/RadixTokenDefinition.js
import BN from 'bn.js';
import { Decimal } from 'decimal.js';
import { RadixUInt256 } from '../..';
import { TSMap } from 'typescript-map';
export var RadixTokenSupplyType;
(function (RadixTokenSupplyType) {
    RadixTokenSupplyType["FIXED"] = "fixed";
    RadixTokenSupplyType["MUTABLE"] = "mutable";
})(RadixTokenSupplyType || (RadixTokenSupplyType = {}));
const NonExpDecimal = Decimal.clone({ toExpPos: 9e15, toExpNeg: -9e15 });
export class RadixTokenDefinition {
    constructor(address, symbol, name, description, granularity = new BN(1), tokenSupplyType, totalSupply, unallocatedTokens, iconUrl) {
        this.totalSupply = new BN(0);
        this.unallocatedTokens = new TSMap();
        this.address = address;
        this.symbol = symbol;
        if (name) {
            this.name = name;
        }
        if (description) {
            this.description = description;
        }
        if (tokenSupplyType) {
            this.tokenSupplyType = tokenSupplyType;
        }
        if (totalSupply) {
            this.totalSupply = totalSupply;
        }
        if (granularity) {
            this.granularity = new RadixUInt256(granularity);
        }
        if (unallocatedTokens) {
            this.unallocatedTokens = unallocatedTokens;
        }
        if (iconUrl) {
            this.iconUrl = iconUrl;
        }
    }
    /**
     * Convert actual decimal token amount to integer subunits stored on the ledger
     * @param amount
     * @returns subunits
     */
    static fromDecimalToSubunits(amount) {
        const inUnits = new NonExpDecimal(amount);
        return new BN(inUnits
            .times(this.SUBUNITS)
            .truncated()
            .toString(), 10);
    }
    /**
     * Convert subunits token amount to actual decimal token amount
     * @param amount
     * @returns token units
     */
    static fromSubunitsToDecimal(amount) {
        const inSubunits = new NonExpDecimal(amount.toString(10));
        return new Decimal(inSubunits.dividedBy(this.SUBUNITS));
    }
    addTotalSupply(difference) {
        this.totalSupply.iadd(new BN(difference));
    }
    getTotalSupply() {
        return this.totalSupply;
    }
    getTokenUnitsTotalSupply() {
        return RadixTokenDefinition.fromSubunitsToDecimal(this.getTotalSupply());
    }
    getGranularity() {
        return this.granularity.value;
    }
    getTokenUnitsGranularity() {
        return RadixTokenDefinition.fromSubunitsToDecimal(this.getGranularity());
    }
    getUnallocatedTokens() {
        return this.unallocatedTokens.values();
    }
    getUnallocatedSupply() {
        const supply = new BN(0);
        for (const unallocatedTokens of this.getUnallocatedTokens()) {
            supply.iadd(unallocatedTokens.getAmount());
        }
        return supply;
    }
}
// All radix tokens are store with 18 subunits
RadixTokenDefinition.SUBUNITS = new Decimal(10).pow(18);
