// radixdlt/build/module/modules/token/RadixTokenManager.js
import { RadixAccount } from '../..';
import { RRI } from '../atommodel';
import { BehaviorSubject, Subject } from 'rxjs';
import { TSMap } from 'typescript-map';
import { filter, timeout, take } from 'rxjs/operators';
import { RadixTokenDefinition } from './RadixTokenDefinition';
/**
 * A singleton class for loading information about tokens
 */
export class RadixTokenManager {
    constructor() {
        this.tokens = {};
        this.tokenSubscriptions = new TSMap();
        this.accounts = new TSMap();
        this.allTokenUpdateSubject = new Subject();
        this.initialized = false;
    }
    initialize(genesis, nativeToken) {
        this.nativeToken = nativeToken;
        const systemAccount = new RadixAccount(nativeToken.address);
        for (const atom of genesis) {
            systemAccount._onAtomReceived({
                action: 'STORE',
                atom,
                processedData: {},
                isHead: false,
            });
        }
        this.accounts.set(systemAccount.getAddress(), systemAccount);
        this.addTokenDefinitionSubscription(nativeToken.toString());
        this.initialized = true;
    }
    async getTokenDefinitionObservable(referenceURI) {
        this.checkInitialized();
        if (!this.tokenSubscriptions.has(referenceURI)) {
            await this.addTokenDefinitionSubscription(referenceURI);
        }
        return this.tokenSubscriptions.get(referenceURI).share();
    }
    async addTokenDefinitionSubscription(referenceURI) {
        const reference = RRI.fromString(referenceURI);
        const account = await this.getAccount(reference.getAddress());
        const placeholderTokenDefinition = new RadixTokenDefinition(reference.getAddress(), reference.getName());
        const bs = new BehaviorSubject(placeholderTokenDefinition);
        account.tokenDefinitionSystem.getTokenDefinitionObservable(reference.getName()).subscribe(bs);
        this.tokenSubscriptions.set(referenceURI, bs);
        bs.subscribe(tokenDefinition => {
            this.tokens[referenceURI] = tokenDefinition;
        });
        bs.subscribe(this.allTokenUpdateSubject);
    }
    getTokenDefinition(referenceURI) {
        this.checkInitialized();
        return new Promise((resolve, reject) => {
            const reference = RRI.fromString(referenceURI);
            this.getAccount(reference.getAddress()).then((account) => {
                account.isSynced().pipe(filter(val => val === true), take(1), timeout(5000)).subscribe(null, error => { reject(new Error('Timeout tying to fetch token information from network')); }, () => {
                    // Account is synced
                    const tokenDefinition = account.tokenDefinitionSystem.getTokenDefinition(reference.getName());
                    if (tokenDefinition) {
                        resolve(tokenDefinition);
                    }
                    else {
                        reject(new Error('Token definition does not exist in the account'));
                    }
                });
            });
        });
    }
    getTokenDefinitionNoLoad(referenceURI) {
        return this.tokens[referenceURI];
    }
    async getAccount(address) {
        if (this.accounts.has(address.toString())) {
            return this.accounts.get(address.toString());
        }
        else {
            const account = new RadixAccount(address);
            await account.openNodeConnection();
            this.accounts.set(address.toString(), account);
            return account;
        }
    }
    checkInitialized() {
        if (!this.initialized) {
            throw new Error('Token Manager not initialized');
        }
    }
    getAllTokenDefinitionUpdates() {
        return this.allTokenUpdateSubject.share();
    }
    /**
     * Return a list of the current tokens in the manager.
     *
     * @returns
     * @memberof RadixToken
     */
    getCurrentTokens() {
        this.checkInitialized();
        return this.tokens;
    }
}
export let radixTokenManager = new RadixTokenManager();
