// radixdlt/build/main/modules/token/RadixTokenManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("../..");
var atommodel_1 = require("../atommodel");
var rxjs_1 = require("rxjs");
var typescript_map_1 = require("typescript-map");
var operators_1 = require("rxjs/operators");
var RadixTokenDefinition_1 = require("./RadixTokenDefinition");
/**
 * A singleton class for loading information about tokens
 */
var RadixTokenManager = /** @class */ (function () {
    function RadixTokenManager() {
        this.tokens = {};
        this.tokenSubscriptions = new typescript_map_1.TSMap();
        this.accounts = new typescript_map_1.TSMap();
        this.allTokenUpdateSubject = new rxjs_1.Subject();
        this.initialized = false;
    }
    RadixTokenManager.prototype.initialize = function (genesis, nativeToken) {
        var e_1, _a;
        this.nativeToken = nativeToken;
        var systemAccount = new __1.RadixAccount(nativeToken.address);
        try {
            for (var genesis_1 = tslib_1.__values(genesis), genesis_1_1 = genesis_1.next(); !genesis_1_1.done; genesis_1_1 = genesis_1.next()) {
                var atom = genesis_1_1.value;
                systemAccount._onAtomReceived({
                    action: 'STORE',
                    atom: atom,
                    processedData: {},
                    isHead: false,
                });
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (genesis_1_1 && !genesis_1_1.done && (_a = genesis_1.return)) _a.call(genesis_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.accounts.set(systemAccount.getAddress(), systemAccount);
        this.addTokenDefinitionSubscription(nativeToken.toString());
        this.initialized = true;
    };
    RadixTokenManager.prototype.getTokenDefinitionObservable = function (referenceURI) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.checkInitialized();
                        if (!!this.tokenSubscriptions.has(referenceURI)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.addTokenDefinitionSubscription(referenceURI)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, this.tokenSubscriptions.get(referenceURI).share()];
                }
            });
        });
    };
    RadixTokenManager.prototype.addTokenDefinitionSubscription = function (referenceURI) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var reference, account, placeholderTokenDefinition, bs;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        reference = atommodel_1.RRI.fromString(referenceURI);
                        return [4 /*yield*/, this.getAccount(reference.getAddress())];
                    case 1:
                        account = _a.sent();
                        placeholderTokenDefinition = new RadixTokenDefinition_1.RadixTokenDefinition(reference.getAddress(), reference.getName());
                        bs = new rxjs_1.BehaviorSubject(placeholderTokenDefinition);
                        account.tokenDefinitionSystem.getTokenDefinitionObservable(reference.getName()).subscribe(bs);
                        this.tokenSubscriptions.set(referenceURI, bs);
                        bs.subscribe(function (tokenDefinition) {
                            _this.tokens[referenceURI] = tokenDefinition;
                        });
                        bs.subscribe(this.allTokenUpdateSubject);
                        return [2 /*return*/];
                }
            });
        });
    };
    RadixTokenManager.prototype.getTokenDefinition = function (referenceURI) {
        var _this = this;
        this.checkInitialized();
        return new Promise(function (resolve, reject) {
            var reference = atommodel_1.RRI.fromString(referenceURI);
            _this.getAccount(reference.getAddress()).then(function (account) {
                account.isSynced().pipe(operators_1.filter(function (val) { return val === true; }), operators_1.take(1), operators_1.timeout(5000)).subscribe(null, function (error) { reject(new Error('Timeout tying to fetch token information from network')); }, function () {
                    // Account is synced
                    var tokenDefinition = account.tokenDefinitionSystem.getTokenDefinition(reference.getName());
                    if (tokenDefinition) {
                        resolve(tokenDefinition);
                    }
                    else {
                        reject(new Error('Token definition does not exist in the account'));
                    }
                });
            });
        });
    };
    RadixTokenManager.prototype.getTokenDefinitionNoLoad = function (referenceURI) {
        return this.tokens[referenceURI];
    };
    RadixTokenManager.prototype.getAccount = function (address) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var account;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.accounts.has(address.toString())) return [3 /*break*/, 1];
                        return [2 /*return*/, this.accounts.get(address.toString())];
                    case 1:
                        account = new __1.RadixAccount(address);
                        return [4 /*yield*/, account.openNodeConnection()];
                    case 2:
                        _a.sent();
                        this.accounts.set(address.toString(), account);
                        return [2 /*return*/, account];
                }
            });
        });
    };
    RadixTokenManager.prototype.checkInitialized = function () {
        if (!this.initialized) {
            throw new Error('Token Manager not initialized');
        }
    };
    RadixTokenManager.prototype.getAllTokenDefinitionUpdates = function () {
        return this.allTokenUpdateSubject.share();
    };
    /**
     * Return a list of the current tokens in the manager.
     *
     * @returns
     * @memberof RadixToken
     */
    RadixTokenManager.prototype.getCurrentTokens = function () {
        this.checkInitialized();
        return this.tokens;
    };
    return RadixTokenManager;
}());
exports.RadixTokenManager = RadixTokenManager;
exports.radixTokenManager = new RadixTokenManager();
