// radixdlt/build/main/modules/token/RadixTokenDefinition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var bn_js_1 = tslib_1.__importDefault(require("bn.js"));
var decimal_js_1 = require("decimal.js");
var __1 = require("../..");
var typescript_map_1 = require("typescript-map");
var RadixTokenSupplyType;
(function (RadixTokenSupplyType) {
    RadixTokenSupplyType["FIXED"] = "fixed";
    RadixTokenSupplyType["MUTABLE"] = "mutable";
})(RadixTokenSupplyType = exports.RadixTokenSupplyType || (exports.RadixTokenSupplyType = {}));
var NonExpDecimal = decimal_js_1.Decimal.clone({ toExpPos: 9e15, toExpNeg: -9e15 });
var RadixTokenDefinition = /** @class */ (function () {
    function RadixTokenDefinition(address, symbol, name, description, granularity, tokenSupplyType, totalSupply, unallocatedTokens, iconUrl) {
        if (granularity === void 0) { granularity = new bn_js_1.default(1); }
        this.totalSupply = new bn_js_1.default(0);
        this.unallocatedTokens = new typescript_map_1.TSMap();
        this.address = address;
        this.symbol = symbol;
        if (name) {
            this.name = name;
        }
        if (description) {
            this.description = description;
        }
        if (tokenSupplyType) {
            this.tokenSupplyType = tokenSupplyType;
        }
        if (totalSupply) {
            this.totalSupply = totalSupply;
        }
        if (granularity) {
            this.granularity = new __1.RadixUInt256(granularity);
        }
        if (unallocatedTokens) {
            this.unallocatedTokens = unallocatedTokens;
        }
        if (iconUrl) {
            this.iconUrl = iconUrl;
        }
    }
    /**
     * Convert actual decimal token amount to integer subunits stored on the ledger
     * @param amount
     * @returns subunits
     */
    RadixTokenDefinition.fromDecimalToSubunits = function (amount) {
        var inUnits = new NonExpDecimal(amount);
        return new bn_js_1.default(inUnits
            .times(this.SUBUNITS)
            .truncated()
            .toString(), 10);
    };
    /**
     * Convert subunits token amount to actual decimal token amount
     * @param amount
     * @returns token units
     */
    RadixTokenDefinition.fromSubunitsToDecimal = function (amount) {
        var inSubunits = new NonExpDecimal(amount.toString(10));
        return new decimal_js_1.Decimal(inSubunits.dividedBy(this.SUBUNITS));
    };
    RadixTokenDefinition.prototype.addTotalSupply = function (difference) {
        this.totalSupply.iadd(new bn_js_1.default(difference));
    };
    RadixTokenDefinition.prototype.getTotalSupply = function () {
        return this.totalSupply;
    };
    RadixTokenDefinition.prototype.getTokenUnitsTotalSupply = function () {
        return RadixTokenDefinition.fromSubunitsToDecimal(this.getTotalSupply());
    };
    RadixTokenDefinition.prototype.getGranularity = function () {
        return this.granularity.value;
    };
    RadixTokenDefinition.prototype.getTokenUnitsGranularity = function () {
        return RadixTokenDefinition.fromSubunitsToDecimal(this.getGranularity());
    };
    RadixTokenDefinition.prototype.getUnallocatedTokens = function () {
        return this.unallocatedTokens.values();
    };
    RadixTokenDefinition.prototype.getUnallocatedSupply = function () {
        var e_1, _a;
        var supply = new bn_js_1.default(0);
        try {
            for (var _b = tslib_1.__values(this.getUnallocatedTokens()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var unallocatedTokens = _c.value;
                supply.iadd(unallocatedTokens.getAmount());
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return supply;
    };
    // All radix tokens are store with 18 subunits
    RadixTokenDefinition.SUBUNITS = new decimal_js_1.Decimal(10).pow(18);
    return RadixTokenDefinition;
}());
exports.RadixTokenDefinition = RadixTokenDefinition;
