// radixdlt/build/main/modules/atommodel/primitives/RadixEUID.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("..");
var long_1 = tslib_1.__importDefault(require("long"));
var id = ':uid:';
var RadixEUID = /** @class */ (function () {
    function RadixEUID(value) {
        if (typeof value === 'number') {
            this.bytes = Buffer.alloc(16);
            this.bytes.writeUInt32BE(value, 12);
        }
        else if (Buffer.isBuffer(value) || Array.isArray(value)) {
            if (value.length === 0) {
                throw new Error('EUID must not be 0 bytes');
            }
            this.bytes = Buffer.from(value);
        }
        else if (typeof value === 'string') {
            this.bytes = Buffer.from(value, 'hex');
        }
        else {
            throw new Error('Unsupported EUID value');
        }
        this.shard = long_1.default.fromBytes(tslib_1.__spread(this.bytes.slice(0, this.bytes.length - 8)));
    }
    RadixEUID.fromJSON = function (data) {
        return new this(data);
    };
    RadixEUID.prototype.toJSON = function () {
        return "" + id + this.bytes.toString('hex');
    };
    RadixEUID.prototype.toDSON = function () {
        return __1.RadixSerializer.toDSON(this);
    };
    RadixEUID.prototype.encodeCBOR = function (encoder) {
        var output = Buffer.alloc(this.bytes.length + 1);
        output.writeInt8(0x02, 0);
        this.bytes.copy(output, 1);
        return encoder.pushAny(output);
    };
    RadixEUID.prototype.equals = function (euid) {
        return this.bytes.compare(euid.bytes) === 0;
    };
    RadixEUID.prototype.toString = function () {
        return this.bytes.toString('hex');
    };
    RadixEUID.prototype.getShard = function () {
        return this.shard;
    };
    RadixEUID = tslib_1.__decorate([
        __1.RadixSerializer.registerPrimitive(id),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], RadixEUID);
    return RadixEUID;
}());
exports.RadixEUID = RadixEUID;
