// radixdlt/build/main/modules/atommodel/primitives/RadixAID.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var long_1 = tslib_1.__importDefault(require("long"));
var __1 = require("..");
var id = ':aid:';
var RadixAID = /** @class */ (function () {
    function RadixAID(bytes) {
        if (bytes.length !== RadixAID_1.BYTES) {
            throw new Error("Bytest lenght must be " + RadixAID_1.BYTES + " but is " + bytes.length);
        }
        this.bytes = Buffer.from(bytes);
    }
    RadixAID_1 = RadixAID;
    RadixAID.from = function (hash, shards) {
        if (shards.length === 0) {
            throw new Error('Shards array cannot be empty');
        }
        var shardIndex = hash[0] % shards.length;
        var selectedShard = shards
            .map(function (s) { return s.toUnsigned(); })
            .sort(function (a, b) { return a.compare(b); })
            .map(function (s) { return s.toSigned(); })[shardIndex];
        var bytes = Buffer.alloc(this.BYTES);
        hash.copy(bytes, 0, 0, this.HASH_BYTES);
        Buffer.from(selectedShard.toBytes()).copy(bytes, this.HASH_BYTES);
        return new this(bytes);
    };
    RadixAID.fromJSON = function (data) {
        return new this(Buffer.from(data, 'hex'));
    };
    RadixAID.prototype.toJSON = function () {
        return "" + id + this.bytes.toString('hex');
    };
    RadixAID.prototype.toDSON = function () {
        return __1.RadixSerializer.toDSON(this);
    };
    RadixAID.prototype.encodeCBOR = function (encoder) {
        var output = Buffer.alloc(this.bytes.length + 1);
        output.writeInt8(0x08, 0);
        this.bytes.copy(output, 1);
        return encoder.pushAny(output);
    };
    RadixAID.prototype.equals = function (aid) {
        return this.bytes.compare(aid.bytes) === 0;
    };
    RadixAID.prototype.toString = function () {
        return this.bytes.toString('hex');
    };
    RadixAID.prototype.getShard = function () {
        return long_1.default.fromBytes(Array.from(this.bytes.subarray(RadixAID_1.HASH_BYTES)));
    };
    var RadixAID_1;
    RadixAID.BYTES = 32;
    RadixAID.HASH_BYTES = 24;
    RadixAID.SHARD_BYTES = 8;
    RadixAID = RadixAID_1 = tslib_1.__decorate([
        __1.RadixSerializer.registerPrimitive(id),
        tslib_1.__metadata("design:paramtypes", [Buffer])
    ], RadixAID);
    return RadixAID;
}());
exports.RadixAID = RadixAID;
