// radixdlt/build/main/modules/atommodel/primitives/RadixBytes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("..");
var RadixBytes = /** @class */ (function () {
    function RadixBytes(bytes) {
        this.bytes = Buffer.from(bytes);
    }
    RadixBytes.fromEncoded = function (encoded) {
        return new this(Buffer.from(encoded, 'base64'));
    };
    RadixBytes.fromJSON = function (bytes) {
        return this.fromEncoded(bytes);
    };
    RadixBytes.prototype.toJSON = function () {
        return ":byt:" + this.bytes.toString('base64');
    };
    RadixBytes.prototype.toDSON = function () {
        return __1.RadixSerializer.toDSON(this);
    };
    RadixBytes.prototype.encodeCBOR = function (encoder) {
        var output = Buffer.alloc(this.bytes.length + 1);
        output.writeInt8(0x01, 0);
        this.bytes.copy(output, 1);
        return encoder.pushAny(output);
    };
    RadixBytes.prototype.toString = function () {
        return this.bytes.toString();
    };
    RadixBytes.prototype.asJSON = function () {
        return JSON.parse(this.bytes.toString('utf8'));
    };
    RadixBytes = tslib_1.__decorate([
        __1.RadixSerializer.registerPrimitive(':byt:'),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], RadixBytes);
    return RadixBytes;
}());
exports.RadixBytes = RadixBytes;
