// radixdlt/build/main/modules/atommodel/primitives/RadixAddress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("..");
var elliptic_1 = tslib_1.__importDefault(require("elliptic"));
var bs58_1 = tslib_1.__importDefault(require("bs58"));
var __2 = require("../../..");
var ec = new elliptic_1.default.ec('secp256k1');
var id = ':adr:';
var RadixAddress = /** @class */ (function () {
    function RadixAddress(magicByte) {
        if (magicByte) {
            this.magicByte = magicByte;
        }
    }
    RadixAddress_1 = RadixAddress;
    RadixAddress.generateNew = function () {
        var radixKeyPair = new RadixAddress_1();
        radixKeyPair.keyPair = ec.genKeyPair();
        return radixKeyPair;
    };
    RadixAddress.fromAddress = function (address) {
        var raw = Array.prototype.slice.call(bs58_1.default.decode(address), 0);
        // Universe check
        if (__2.radixUniverse && __2.radixUniverse.initialized && __2.radixUniverse.getMagicByte() !== raw[0]) {
            throw new Error('Address is from a different universe');
        }
        // Checksum
        var check = __2.radixHash(raw.splice(0, raw.length - 4), 0, raw.length - 4);
        for (var i = 0; i < 4; i++) {
            if (check[i] !== raw[raw.length - 4 + i]) {
                throw new Error('Invalid address');
            }
        }
        raw = Array.prototype.slice.call(bs58_1.default.decode(address), 0);
        var radixAddress = new this(raw[0]);
        radixAddress.keyPair = ec.keyFromPublic(raw.splice(1, raw.length - 5));
        return radixAddress;
    };
    RadixAddress.fromPublic = function (publicKey, magicByte) {
        if (!publicKey) {
            throw new Error('Missing public key');
        }
        if (publicKey.length !== 33) {
            throw new Error('Public key must be 33 bytes, but was ' + publicKey.length);
        }
        var radixAddress = new this(magicByte);
        radixAddress.keyPair = ec.keyFromPublic(publicKey);
        return radixAddress;
    };
    RadixAddress.fromPrivate = function (privateKey, magicByte) {
        var radixAddress = new this(magicByte);
        radixAddress.keyPair = ec.keyFromPrivate(privateKey);
        return radixAddress;
    };
    RadixAddress.prototype.getAddressBytes = function () {
        var publicKey = this.keyPair.getPublic().encode('array', true);
        var addressBytes = [];
        addressBytes[0] = this.magicByte ? this.magicByte : __2.radixUniverse.getMagicByte();
        for (var i = 0; i < publicKey.length; i++) {
            addressBytes[i + 1] = publicKey[i];
        }
        var check = __2.radixHash(addressBytes, 0, publicKey.length + 1);
        for (var i = 0; i < 4; i++) {
            addressBytes[publicKey.length + 1 + i] = check[i];
        }
        return Buffer.from(addressBytes);
    };
    RadixAddress.prototype.getAddress = function () {
        return bs58_1.default.encode(this.getAddressBytes());
    };
    RadixAddress.prototype.getHash = function () {
        return __2.radixHash(this.getPublic(), 0, this.getPublic().length);
    };
    RadixAddress.prototype.getUID = function () {
        var hash = this.getHash();
        return new __1.RadixEUID(hash.slice(0, 16));
    };
    RadixAddress.prototype.getShard = function () {
        return this.getUID().getShard();
    };
    RadixAddress.prototype.getPublic = function () {
        return Buffer.from(this.keyPair.getPublic(true, 'array'));
    };
    RadixAddress.prototype.getPrivate = function () {
        return this.keyPair.getPrivate().toArrayLike(Buffer);
    };
    RadixAddress.prototype.sign = function (data) {
        var signature = this.keyPair.sign(data);
        return __1.RadixECSignature.fromEllasticSignature(signature);
    };
    RadixAddress.prototype.verify = function (data, signature) {
        return this.keyPair.verify(data, { r: signature.r.bytes, s: signature.s.bytes });
    };
    RadixAddress.prototype.equals = function (otherAddress) {
        return this.getAddress() === otherAddress.getAddress();
    };
    RadixAddress.prototype.toString = function () {
        return this.getAddress();
    };
    RadixAddress.fromJSON = function (address) {
        return this.fromAddress(address);
    };
    RadixAddress.prototype.toJSON = function () {
        return "" + id + this.getAddress();
    };
    RadixAddress.prototype.toDSON = function () {
        return __1.RadixSerializer.toDSON(this);
    };
    RadixAddress.prototype.encodeCBOR = function (encoder) {
        var addressBuffer = this.getAddressBytes();
        var output = Buffer.alloc(addressBuffer.length + 1);
        output.writeInt8(0x04, 0);
        addressBuffer.copy(output, 1);
        return encoder.pushAny(output);
    };
    var RadixAddress_1;
    RadixAddress = RadixAddress_1 = tslib_1.__decorate([
        __1.RadixSerializer.registerPrimitive(id),
        tslib_1.__metadata("design:paramtypes", [Number])
    ], RadixAddress);
    return RadixAddress;
}());
exports.RadixAddress = RadixAddress;
