// radixdlt/build/main/modules/atommodel/primitives/RRI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("..");
var id = ':rri:';
var RRI = /** @class */ (function () {
    function RRI(address, name) {
        this.address = address;
        this.name = name;
    }
    RRI.fromJSON = function (uri) {
        return this.fromString(uri);
    };
    RRI.fromString = function (uri) {
        var parts = uri.split('/');
        if (parts.length !== 3) {
            throw new Error('RRI must be of the format /:address/:unique');
        }
        return new this(__1.RadixAddress.fromAddress(parts[1]), parts[2]);
    };
    RRI.prototype.toJSON = function () {
        return "" + id + this.toString();
    };
    RRI.prototype.toString = function () {
        return "/" + this.address.toString() + "/" + this.name;
    };
    RRI.prototype.toDSON = function () {
        return __1.RadixSerializer.toDSON(this);
    };
    RRI.prototype.encodeCBOR = function (encoder) {
        var s = Buffer.from(this.toString(), 'utf8');
        var output = Buffer.alloc(s.length + 1);
        output.writeInt8(0x06, 0);
        s.copy(output, 1);
        return encoder.pushAny(output);
    };
    RRI.prototype.equals = function (rri) {
        return this.address.equals(rri.address) && this.name === rri.name;
    };
    RRI.prototype.getAddress = function () {
        return this.address;
    };
    RRI.prototype.getName = function () {
        return this.name;
    };
    RRI = tslib_1.__decorate([
        __1.RadixSerializer.registerPrimitive(id),
        tslib_1.__metadata("design:paramtypes", [__1.RadixAddress, String])
    ], RRI);
    return RRI;
}());
exports.RRI = RRI;
