// radixdlt/build/main/modules/cache/RadixNEDBAtomCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var nedb_1 = tslib_1.__importDefault(require("nedb"));
var atommodel_1 = require("../atommodel");
var RadixLogger_1 = require("../common/RadixLogger");
var RadixNEDBAtomCache = /** @class */ (function () {
    /**
     * Creates an instance of radix nedbatom cache.
     * @param filename File path on disk in Node, path in localStorage in the browser
     */
    function RadixNEDBAtomCache(filename) {
        var _this = this;
        this.storeAtom = function (atom) {
            return _this.notExists({ _id: atom.getAidString() })
                .then(function () {
                // Serialize
                var serializedAtom = atom.toJSON();
                serializedAtom['_id'] = atom.getAidString();
                // Store
                return _this.insert(serializedAtom);
            })
                .then(function (newDoc) {
                return atom;
            })
                .catch(function (error) {
                RadixLogger_1.logger.trace('Atom already in DB');
            });
        };
        this.getAtoms = function (queryAddress) {
            // Find
            var query = {};
            // logger.info(query)
            return _this.find(query).then(function (atoms) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                var deserialized, atoms_1, atoms_1_1, atom, deserializedAtom, atomAddresses, atomAddresses_1, atomAddresses_1_1, address, e_1_1;
                var e_1, _a, e_2, _b;
                return tslib_1.__generator(this, function (_c) {
                    switch (_c.label) {
                        case 0:
                            deserialized = [];
                            _c.label = 1;
                        case 1:
                            _c.trys.push([1, 6, 7, 8]);
                            atoms_1 = tslib_1.__values(atoms), atoms_1_1 = atoms_1.next();
                            _c.label = 2;
                        case 2:
                            if (!!atoms_1_1.done) return [3 /*break*/, 5];
                            atom = atoms_1_1.value;
                            return [4 /*yield*/, this.asyncDeserialize(atom)];
                        case 3:
                            deserializedAtom = _c.sent();
                            atomAddresses = deserializedAtom.getAddresses();
                            try {
                                for (atomAddresses_1 = (e_2 = void 0, tslib_1.__values(atomAddresses)), atomAddresses_1_1 = atomAddresses_1.next(); !atomAddresses_1_1.done; atomAddresses_1_1 = atomAddresses_1.next()) {
                                    address = atomAddresses_1_1.value;
                                    if (address.equals(queryAddress)) {
                                        deserialized.push(deserializedAtom);
                                        break;
                                    }
                                }
                            }
                            catch (e_2_1) { e_2 = { error: e_2_1 }; }
                            finally {
                                try {
                                    if (atomAddresses_1_1 && !atomAddresses_1_1.done && (_b = atomAddresses_1.return)) _b.call(atomAddresses_1);
                                }
                                finally { if (e_2) throw e_2.error; }
                            }
                            _c.label = 4;
                        case 4:
                            atoms_1_1 = atoms_1.next();
                            return [3 /*break*/, 2];
                        case 5: return [3 /*break*/, 8];
                        case 6:
                            e_1_1 = _c.sent();
                            e_1 = { error: e_1_1 };
                            return [3 /*break*/, 8];
                        case 7:
                            try {
                                if (atoms_1_1 && !atoms_1_1.done && (_a = atoms_1.return)) _a.call(atoms_1);
                            }
                            finally { if (e_1) throw e_1.error; }
                            return [7 /*endfinally*/];
                        case 8: return [2 /*return*/, deserialized];
                    }
                });
            }); });
        };
        // Promise wrappers for nedb
        this.findOne = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.findOne(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.notExists = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.findOne(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else if (!doc) {
                        resolve(true);
                    }
                    reject('Atom already in db');
                });
            });
        };
        this.find = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.find(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.insert = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.insert(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.remove = function (opt) {
            return new Promise(function (resolve, reject) {
                _this.db.remove(opt, function (error, doc) {
                    if (error) {
                        reject(error);
                    }
                    else {
                        resolve(doc);
                    }
                });
            });
        };
        this.db = new nedb_1.default({
            filename: filename,
            autoload: true,
        });
    }
    /**
     * Clears all atoms from the cache
     */
    RadixNEDBAtomCache.prototype.reset = function () {
        this.db.remove({}, { multi: true }, function (error, numRemoved) {
            // Do nothing
        });
    };
    RadixNEDBAtomCache.prototype.asyncDeserialize = function (atom) {
        return new Promise(function (resolve, reject) {
            setTimeout(function () {
                delete atom['_id'];
                resolve(atommodel_1.RadixSerializer.fromJSON(atom));
            }, 0);
        });
    };
    RadixNEDBAtomCache.prototype.deleteAtom = function (atom) {
        return this.remove({ _id: atom.getAidString() });
    };
    return RadixNEDBAtomCache;
}());
exports.default = RadixNEDBAtomCache;
