// radixdlt/build/main/modules/cache/RadixCacheAccountSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var RadixCacheAccountSystem = /** @class */ (function () {
    function RadixCacheAccountSystem(address, atomCache) {
        this.address = address;
        this.name = 'CACHE';
        if (atomCache) {
            this.atomCache = atomCache;
        }
    }
    RadixCacheAccountSystem.prototype.processAtomUpdate = function (atomUpdate) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                if (!this.atomCache) {
                    return [2 /*return*/];
                }
                // Just put it in the cache
                if (atomUpdate.action === 'STORE') {
                    this.atomCache.storeAtom(atomUpdate.atom);
                }
                else if (atomUpdate.action === 'DELETE') {
                    this.atomCache.deleteAtom(atomUpdate.atom);
                }
                return [2 /*return*/];
            });
        });
    };
    RadixCacheAccountSystem.prototype.loadAtoms = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, this.atomCache.getAtoms(this.address)];
            });
        });
    };
    return RadixCacheAccountSystem;
}());
exports.default = RadixCacheAccountSystem;
