// radixdlt/build/module/modules/universe/RadixUniverseConfig.js
import betanet from './configs/betanet.json';
import sunstone from './configs/sunstone.json';
import local from './configs/local.json';
import Long from 'long';
import { RadixSerializer } from '../atommodel';
export default class RadixUniverseConfig {
    constructor(rawJson) {
        this.rawJson = rawJson;
        this.magic = rawJson.magic;
        this.magicByte = Long.fromNumber(this.magic).and(0xff).toNumber();
    }
    initialize() {
        const obj = RadixSerializer.fromJSON(this.rawJson);
        this.port = obj.port;
        this.name = obj.name;
        this.description = obj.description;
        this.type = obj.type;
        this.timestamp = obj.timestamp;
        this.creator = obj.creator;
        this.genesis = obj.genesis;
    }
    getMagic() {
        return this.magic;
    }
    getMagicByte() {
        return this.magicByte;
    }
}
RadixUniverseConfig.SUNSTONE = new RadixUniverseConfig(sunstone);
RadixUniverseConfig.LOCAL = new RadixUniverseConfig(local);
RadixUniverseConfig.BETANET = new RadixUniverseConfig(betanet);
