// radixdlt/build/main/modules/messaging/RadixMessagingAccountSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var rxjs_1 = require("rxjs");
var typescript_map_1 = require("typescript-map");
var RadixDecryptionAccountSystem_1 = require("../account/RadixDecryptionAccountSystem");
var RadixMessagingAccountSystem = /** @class */ (function () {
    function RadixMessagingAccountSystem(address) {
        this.address = address;
        this.name = 'RADIX-MESSAGING';
        this.messageSubject = new rxjs_1.Subject();
        this.chats = new typescript_map_1.TSMap();
        this.messages = new typescript_map_1.TSMap();
    }
    RadixMessagingAccountSystem.prototype.processAtomUpdate = function (atomUpdate) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                if (!('decryptedData' in atomUpdate.processedData) ||
                    atomUpdate.processedData.decryptedData.application !== 'message' ||
                    atomUpdate.processedData.decryptedData.decryptionState === RadixDecryptionAccountSystem_1.RadixDecryptionState.CANNOT_DECRYPT) {
                    return [2 /*return*/];
                }
                if (atomUpdate.action === 'STORE') {
                    this.processStoreAtom(atomUpdate);
                }
                else if (atomUpdate.action === 'DELETE') {
                    this.processDeleteAtom(atomUpdate);
                }
                return [2 /*return*/];
            });
        });
    };
    RadixMessagingAccountSystem.prototype.startNewChat = function (to) {
        // Create new chat
        var chatId = to.getAddress();
        if (this.chats.has(chatId)) {
            return;
        }
        var chatDescription = {
            address: to.getAddress(),
            chat_id: chatId,
            title: chatId,
            last_message_timestamp: Date.now(),
            messages: new typescript_map_1.TSMap(),
        };
        // Add at the top
        this.chats.set(chatId, chatDescription);
    };
    RadixMessagingAccountSystem.prototype.processStoreAtom = function (atomUpdate) {
        var atom = atomUpdate.atom;
        var aid = atom.getAidString();
        var signatures = atom.signatures;
        var decryptedData = atomUpdate.processedData.decryptedData;
        // Skip existing atoms
        if (this.messages.has(aid)) {
            return;
        }
        var from = decryptedData.from;
        var to = decryptedData.to.find(function (a) { return !a.equals(from); });
        if (!to) {
            throw new Error('A message needs to have at least one other recipient');
        }
        // Chat id
        var address = null;
        var isMine = false;
        if (from.equals(this.address)) {
            address = to;
            isMine = true;
        }
        else {
            address = from;
        }
        if (address === null) {
            throw new Error('Error processing a radix-message atom: neither of addresses is owned by this account');
        }
        var chatId = address.toString();
        var message = {
            aid: aid,
            chat_id: chatId,
            to: to,
            from: from,
            content: decryptedData.data,
            timestamp: atom.getTimestamp(),
            is_mine: isMine,
            encryptionState: decryptedData.decryptionState,
        };
        // Find existing chat
        // Otherwise create new chat
        if (!this.chats.has(chatId)) {
            var newChatDescription = {
                address: address.getAddress(),
                chat_id: chatId,
                title: chatId,
                last_message_timestamp: message.timestamp,
                messages: new typescript_map_1.TSMap(),
            };
            this.chats.set(chatId, newChatDescription);
        }
        var chatDescription = this.chats.get(chatId);
        if (message.timestamp > chatDescription.last_message_timestamp) {
            chatDescription.last_message_timestamp = message.timestamp;
        }
        chatDescription.messages.set(aid, message);
        // Move chat to the top
        this.chats.delete(chatId);
        this.chats.set(chatId, chatDescription);
        this.messages.set(aid, message);
        var messageUpdate = {
            action: 'STORE',
            aid: aid,
            message: message,
        };
        this.messageSubject.next(messageUpdate);
    };
    RadixMessagingAccountSystem.prototype.processDeleteAtom = function (atomUpdate) {
        var atom = atomUpdate.atom;
        var aid = atom.getAidString();
        // Skip nonexisting atoms
        if (!this.messages.has(aid)) {
            return;
        }
        var message = this.messages.get(aid);
        this.chats.get(message.chat_id).messages.delete(aid);
        this.messages.delete(aid);
        var messageUpdate = {
            action: 'DELETE',
            aid: aid,
            message: message,
        };
        this.messageSubject.next(messageUpdate);
    };
    RadixMessagingAccountSystem.prototype.getAllMessages = function () {
        var _this = this;
        return rxjs_1.Observable.create(function (observer) {
            var e_1, _a;
            try {
                // Send all old transactions
                for (var _b = tslib_1.__values(_this.messages.values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var message = _c.value;
                    var messageUpdate = {
                        action: 'STORE',
                        aid: message.aid,
                        message: message,
                    };
                    observer.next(messageUpdate);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            // Subscribe for new ones
            _this.messageSubject.subscribe(observer);
        });
    };
    return RadixMessagingAccountSystem;
}());
exports.default = RadixMessagingAccountSystem;
