// radixdlt/build/main/modules/atommodel/serializer/RadixSerializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var cbor_1 = tslib_1.__importDefault(require("cbor"));
var typescript_map_1 = require("typescript-map");
var RadixLogger_1 = require("../../common/RadixLogger");
require("reflect-metadata");
var RadixUtil_1 = require("../../common/RadixUtil");
exports.JSON_PROPERTIES_KEY = 'JSON_SERIALIZATION_PROPERTIES';
exports.DSON_PROPERTIES_KEY = 'DSON_SERIALIZATION_PROPERTIES';
/**
 * Decorator to register a property for JSON serialization.
 *   Stores property names in the prototype metadata
 * @param target
 * @param propertyName
 */
function includeJSON(target, propertyName) {
    registerPropertyForSerialization(target, propertyName, exports.JSON_PROPERTIES_KEY);
}
exports.includeJSON = includeJSON;
/**
 * Decorator to register a property for DSON serialization.
 *   Stores property names in the prototype metadata
 * @param target
 * @param propertyName
 */
function includeDSON(target, propertyName) {
    registerPropertyForSerialization(target, propertyName, exports.DSON_PROPERTIES_KEY);
}
exports.includeDSON = includeDSON;
/**
 * Registers property for serialization.
 *   Stores an array of property names in the prototype metadata.
 *   Inherits from the object prototype chain.
 * @param target
 * @param propertyName
 * @param metadataKey
 */
function registerPropertyForSerialization(target, propertyName, metadataKey) {
    // Avoid modifying parent metadata, clone instead
    if (!Reflect.hasOwnMetadata(metadataKey, target)) {
        var props = [];
        if (Reflect.hasMetadata(metadataKey, target)) {
            // Clone parent metadata
            props = Reflect.getMetadata(metadataKey, target).slice(0);
        }
        Reflect.defineMetadata(metadataKey, props, target);
    }
    Reflect.getMetadata(metadataKey, target).push(propertyName);
    Reflect.getMetadata(metadataKey, target).sort();
}
var RadixSerializer = /** @class */ (function () {
    function RadixSerializer() {
    }
    /**
     * Decorator to register a class for serialization
     * @param serializer serializer id
     * @returns
     */
    RadixSerializer.registerClass = function (serializer) {
        var _this = this;
        return function (constructor) {
            constructor.SERIALIZER = serializer;
            _this.classes.set(serializer, constructor);
        };
    };
    /**
     * Registers advanced 'primitive' types for serialization
     * @param tag
     * @returns
     */
    RadixSerializer.registerPrimitive = function (tag) {
        var _this = this;
        return function (constructor) {
            _this.primitives.set(tag, constructor);
        };
    };
    RadixSerializer.toJSON = function (data) {
        var e_1, _a;
        if (Array.isArray(data)) {
            var output = [];
            try {
                for (var data_1 = tslib_1.__values(data), data_1_1 = data_1.next(); !data_1_1.done; data_1_1 = data_1.next()) {
                    var item = data_1_1.value;
                    output.push(RadixSerializer.toJSON(item));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (data_1_1 && !data_1_1.done && (_a = data_1.return)) _a.call(data_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return output;
        }
        else if (typeof data === 'number' ||
            typeof data === 'boolean') {
            return data;
        }
        else if (typeof data === 'string') {
            return ":str:" + data;
        }
        else if (data !== null && typeof data === 'object') {
            if (typeof data.toJSON === 'function') {
                return data.toJSON();
            }
            else {
                var output = {};
                for (var key in data) {
                    var serialized = RadixSerializer.toJSON(data[key]);
                    if (!RadixUtil_1.isEmpty(serialized)) {
                        output[key] = serialized;
                    }
                }
                return output;
            }
        }
    };
    RadixSerializer.fromJSON = function (json) {
        var e_2, _a;
        if (Array.isArray(json)) {
            var output = [];
            try {
                for (var json_1 = tslib_1.__values(json), json_1_1 = json_1.next(); !json_1_1.done; json_1_1 = json_1.next()) {
                    var item = json_1_1.value;
                    output.push(RadixSerializer.fromJSON(item));
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (json_1_1 && !json_1_1.done && (_a = json_1.return)) _a.call(json_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return output;
        }
        else if (typeof json === 'object') {
            return RadixSerializer.fromJSONObject(json);
        }
        else if (typeof json === 'string') {
            // Advanced primitives
            var tag = json.slice(0, 5);
            if (tag === ':str:') {
                return json.slice(5);
            }
            // Cast to a primitve class
            if (this.primitives.has(tag)) {
                return this.primitives.get(tag).fromJSON(json.slice(5));
            }
            RadixLogger_1.logger.warn("No matching class for primitive string \"" + json + "\"");
        }
        else {
            return json;
        }
    };
    RadixSerializer.fromJSONObject = function (jsonObject) {
        var output = {};
        for (var key in jsonObject) {
            output[key] = (key === 'serializer' ? jsonObject[key] : RadixSerializer.fromJSON(jsonObject[key]));
        }
        if ('serializer' in output) {
            // tslint:disable-next-line:no-string-literal
            var type = output['serializer'];
            if (this.classes.has(type)) {
                return this.classes.get(type).fromJSON(output);
            }
            RadixLogger_1.logger.warn("Serializer \"" + type + "\" not implemented");
        }
        return output;
    };
    RadixSerializer.toDSON = function (data) {
        var enc = new cbor_1.default.Encoder();
        // Overide default object encoder to use stream encoding and lexicographical ordering of keys
        enc.addSemanticType(Object, function (encoder, obj) {
            var e_3, _a;
            var keys = Object.keys(obj);
            keys.sort();
            if (!encoder.push(Buffer.from([191]))) {
                return false;
            }
            try {
                for (var keys_1 = tslib_1.__values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                    var key = keys_1_1.value;
                    if (RadixUtil_1.isEmpty(obj[key])) {
                        continue;
                    }
                    if (!encoder.pushAny(key)) {
                        return false;
                    }
                    if (!encoder.pushAny(obj[key])) {
                        return false;
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
                }
                finally { if (e_3) throw e_3.error; }
            }
            if (!encoder.push(Buffer.from([0xFF]))) {
                return false;
            }
            return true;
        });
        return enc._encodeAll([data]);
    };
    RadixSerializer.classes = new typescript_map_1.TSMap();
    RadixSerializer.primitives = new typescript_map_1.TSMap();
    return RadixSerializer;
}());
exports.RadixSerializer = RadixSerializer;
