// radixdlt/build/main/modules/identity/RadixRemoteIdentity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var RadixIdentity_1 = tslib_1.__importDefault(require("./RadixIdentity"));
var atommodel_1 = require("../atommodel");
var RadixLogger_1 = require("../common/RadixLogger");
var rpc_websockets_1 = require("rpc-websockets");
var RadixRemoteIdentity = /** @class */ (function (_super) {
    tslib_1.__extends(RadixRemoteIdentity, _super);
    function RadixRemoteIdentity(address, token, remoteUrl) {
        var _this = _super.call(this, address) || this;
        _this.address = address;
        _this.token = token;
        _this.remoteUrl = remoteUrl;
        return _this;
    }
    /**
     * Checks if a given web socket connection is still alive, otherwise creates a new one
     *
     * @returns A WebSocket connection
     */
    RadixRemoteIdentity.prototype.getSocketConnection = function () {
        this.socket = new rpc_websockets_1.Client(this.remoteUrl);
        this.socket.on('error', function (error) { return RadixLogger_1.logger.error(error); });
        this.socket.on('close', function () { return RadixLogger_1.logger.info('Socket closed'); });
        return this.socket;
    };
    /**
     * Signs an atom with the wallet using the remote identity
     *
     * @param atom - The atom to be signed
     * @returns A promise with the signed atom
     */
    RadixRemoteIdentity.prototype.signAtom = function (atom) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var socket = _this.getSocketConnection();
            socket.on('open', function () {
                socket.call('sign_atom', {
                    token: _this.token,
                    atom: atommodel_1.RadixSerializer.toJSON(atom),
                }).then(function (response) {
                    atom.signatures = atommodel_1.RadixSerializer.fromJSON(response);
                    resolve(atom);
                }).catch(function (error) {
                    reject(error);
                }).finally(function () {
                    socket.close();
                });
            });
        });
    };
    /**
     * Decrypt the payload of an atom
     *
     * @param payload - The payload of the atom to be decrypted
     * @returns A promise with the decrypted payload
     */
    RadixRemoteIdentity.prototype.decryptECIESPayload = function (payload) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        var socket = _this.getSocketConnection();
                        socket.on('open', function () {
                            socket.call('decrypt_ecies_payload', {
                                token: _this.token,
                                payload: payload,
                            })
                                .then(function (response) {
                                resolve(Buffer.from(response.data));
                            })
                                .catch(function (error) {
                                reject(error);
                            }).finally(function () {
                                socket.close();
                            });
                        });
                    })];
            });
        });
    };
    RadixRemoteIdentity.prototype.decryptECIESPayloadWithProtectors = function (protectors, payload) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        var socket = _this.getSocketConnection();
                        socket.on('open', function () {
                            socket.call('decrypt_ecies_payload_with_protectors', {
                                token: _this.token,
                                protectors: protectors,
                                payload: payload,
                            })
                                .then(function (response) {
                                resolve(Buffer.from(response.data));
                            })
                                .catch(function (error) {
                                reject(error);
                            }).finally(function () {
                                socket.close();
                            });
                        });
                    })];
            });
        });
    };
    /**
     * Returns the public key of this identity synchronously
     *
     * @returns The public key of the identity
     */
    RadixRemoteIdentity.prototype.getPublicKey = function () {
        return this.address.getPublic();
    };
    /**
     * Creates a new instance of a RadixRemoteIdentity
     *
     * @param name - The name of the application that wants to use the remote identity
     * @param description - The description of the application that wants to use the remote identity
     * @param [host] - The host of the wallet
     * @param [port] - The port in which the wallet server is being exposed
     * @returns A promise with an instance of a RadixRemoteIdentity
     */
    RadixRemoteIdentity.createNew = function (name, description, permissions, host, port) {
        if (permissions === void 0) { permissions = ['sign_atom', 'decrypt_ecies_payload', 'get_public_key']; }
        if (host === void 0) { host = 'localhost'; }
        if (port === void 0) { port = '54345'; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var token, publicKey, error_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, RadixRemoteIdentity.register(name, description, permissions, host, port)];
                    case 1:
                        token = _a.sent();
                        return [4 /*yield*/, RadixRemoteIdentity.getRemotePublicKey(token, host, port)];
                    case 2:
                        publicKey = _a.sent();
                        return [2 /*return*/, new RadixRemoteIdentity(atommodel_1.RadixAddress.fromPublic(publicKey), token, "ws://" + host + ":" + port)];
                    case 3:
                        error_1 = _a.sent();
                        throw error_1;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Registers a RadixRemoteIdentity to the wallet
     *
     * @param name - The name of the application that wants to use the remote identity
     * @param description - The description of the application that wants to use the remote identity
     * @param [host] - The host of the wallet
     * @param [port] - The port in which the wallet server is being exposed
     * @returns A promise with a valid token to interact with the wallet
     */
    RadixRemoteIdentity.register = function (name, description, permissions, host, port) {
        if (host === void 0) { host = 'localhost'; }
        if (port === void 0) { port = '54345'; }
        return new Promise(function (resolve, reject) {
            // This is an independant websocket because 'register' is a static method
            var socket = new rpc_websockets_1.Client("ws://" + host + ":" + port);
            socket.on('open', function () {
                socket.call('register', {
                    name: name,
                    description: description,
                    permissions: permissions,
                }).then(function (response) {
                    resolve(response.token);
                }).catch(function (error) {
                    reject(error);
                });
            });
        });
    };
    /**
     * Returns the public key of this identity asynchronously
     *
     * @param [host] - The host of the wallet
     * @param [port] - The port in which the wallet server is being exposed
     * @returns A promise with the public key of the identity
     */
    RadixRemoteIdentity.getRemotePublicKey = function (token, host, port) {
        if (host === void 0) { host = 'localhost'; }
        if (port === void 0) { port = '54345'; }
        return new Promise(function (resolve, reject) {
            // This is an independant websocket because 'getRemotePublicKey' is a static method
            var socket = new rpc_websockets_1.Client("ws://" + host + ":" + port);
            socket.on('open', function () {
                socket.call('get_public_key', {
                    token: token,
                }).then(function (response) {
                    resolve(response.data);
                }).catch(function (error) {
                    reject(error);
                }).finally(function () {
                    socket.close();
                });
            });
        });
    };
    /**
     * Determines whether the server is up or down
     *
     * @param [host] - The host of the wallet
     * @param [port] - The port in which the wallet server is being exposed
     * @returns A promise with true or false whether the server is up or down
     */
    RadixRemoteIdentity.isServerUp = function (host, port) {
        if (host === void 0) { host = 'localhost'; }
        if (port === void 0) { port = '54345'; }
        return new Promise(function (resolve, reject) {
            var socket = new rpc_websockets_1.Client("ws://" + host + ":" + port);
            socket.on('open', function () { return resolve(true); });
            setTimeout(function () {
                if (socket && socket.ready) {
                    socket.close();
                    resolve(true);
                }
                else {
                    socket.close();
                    resolve(false);
                }
            }, 2000);
        });
    };
    return RadixRemoteIdentity;
}(RadixIdentity_1.default));
exports.default = RadixRemoteIdentity;
