// list/dist/es/ramda.js
import { curry } from "ramda";
import * as L from "./index";
export { Node, List, list } from "./index";
export var prepend = curry(L.prepend);
export var isList = curry(L.isList);
export var append = curry(L.append);
export var of = curry(L.of);
export var pair = curry(L.pair);
export var empty = curry(L.empty);
export var repeat = curry(L.repeat);
export var times = curry(L.times);
export var length = curry(L.length);
export var first = curry(L.first);
export var head = first;
export var last = curry(L.last);
export var nth = curry(L.nth);
export var map = curry(L.map);
export var forEach = curry(L.forEach);
export var pluck = curry(L.pluck);
export var range = curry(L.range);
export var foldl = curry(L.foldl);
export var reduce = foldl;
export var filter = curry(L.filter);
export var reject = curry(L.reject);
export var partition = curry(L.partition);
export var join = curry(L.join);
export var foldr = curry(L.foldr);
export var reduceRight = foldr;
export var ap = curry(L.ap);
export var chain = curry(L.chain);
export var flatten = curry(L.flatten);
export var every = curry(L.every);
export var all = every;
export var some = curry(L.some);
export var any = some;
export var none = curry(L.none);
export var find = curry(L.find);
export var indexOf = curry(L.indexOf);
export var findIndex = curry(L.findIndex);
export var includes = curry(L.includes);
export var contains = includes;
export var equals = curry(L.equals);
export var concat = curry(L.concat);
export var update = curry(L.update);
export var adjust = curry(L.adjust);
export var slice = curry(L.slice);
export var take = curry(L.take);
export var takeWhile = curry(L.takeWhile);
export var dropWhile = curry(L.dropWhile);
export var takeLast = curry(L.takeLast);
export var splitAt = curry(L.splitAt);
export var remove = curry(L.remove);
export var reverse = curry(L.reverse);
export var drop = curry(L.drop);
export var dropLast = curry(L.dropLast);
export var pop = curry(L.pop);
export var init = pop;
export var tail = curry(L.tail);
export var toArray = curry(L.toArray);
export var from = curry(L.from);
export var insert = curry(L.insert);
export var insertAll = curry(L.insertAll);
export var zip = curry(L.zip);
export var zipWith = curry(L.zipWith);
export var sort = curry(L.sort);
export var sortWith = curry(L.sortWith);
export var sortBy = curry(L.sortBy);
