// radixdlt/build/main/modules/atommodel/RadixSerializableObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var _1 = require(".");
var RadixUtil_1 = require("../common/RadixUtil");
var RadixSerializableObject = /** @class */ (function () {
    function RadixSerializableObject() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.version = 100;
        // Empty constructor
    }
    RadixSerializableObject.fromJSON = function (json) {
        // So that we can have constructors for the different classes
        var obj = Object.create(this.prototype);
        if (json) {
            for (var key in json) {
                if (key === 'constructor' || key === 'serializationProperties') {
                    continue;
                }
                obj[key] = json[key];
            }
        }
        return obj;
    };
    Object.defineProperty(RadixSerializableObject.prototype, "serializer", {
        get: function () {
            return this.constructor.SERIALIZER;
        },
        set: function (_) {
            // Do nothing
        },
        enumerable: true,
        configurable: true
    });
    RadixSerializableObject.prototype.toJSON = function () {
        var e_1, _a;
        var constructor = this.constructor;
        var output = { serializer: constructor.SERIALIZER };
        var serializationProps = Reflect.getMetadata(_1.JSON_PROPERTIES_KEY, this);
        try {
            for (var serializationProps_1 = tslib_1.__values(serializationProps), serializationProps_1_1 = serializationProps_1.next(); !serializationProps_1_1.done; serializationProps_1_1 = serializationProps_1.next()) {
                var key = serializationProps_1_1.value;
                var serialized = _1.RadixSerializer.toJSON(this[key]);
                if (!RadixUtil_1.isEmpty(serialized)) {
                    output[key] = serialized;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (serializationProps_1_1 && !serializationProps_1_1.done && (_a = serializationProps_1.return)) _a.call(serializationProps_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return output;
    };
    RadixSerializableObject.prototype.toDSON = function () {
        return _1.RadixSerializer.toDSON(this);
    };
    RadixSerializableObject.prototype.encodeCBOR = function (encoder) {
        var e_2, _a;
        // Streaming encoding for maps
        var serializationProps = Reflect.getMetadata(_1.DSON_PROPERTIES_KEY, this);
        if (!encoder.push(Buffer.from([191]))) {
            return false;
        }
        try {
            for (var serializationProps_2 = tslib_1.__values(serializationProps), serializationProps_2_1 = serializationProps_2.next(); !serializationProps_2_1.done; serializationProps_2_1 = serializationProps_2.next()) {
                var prop = serializationProps_2_1.value;
                if (RadixUtil_1.isEmpty(this[prop])) {
                    continue;
                }
                if (!encoder.pushAny(prop)) {
                    return false;
                }
                if (!encoder.pushAny(this[prop])) {
                    return false;
                }
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (serializationProps_2_1 && !serializationProps_2_1.done && (_a = serializationProps_2.return)) _a.call(serializationProps_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
        if (!encoder.push(Buffer.from([0xFF]))) {
            return false;
        }
        return true;
    };
    RadixSerializableObject.prototype.getHash = function () {
        return RadixUtil_1.radixHash(this.toDSON());
    };
    RadixSerializableObject.prototype.getSize = function () {
        return this.toDSON().length;
    };
    RadixSerializableObject.SERIALIZER = '';
    tslib_1.__decorate([
        _1.includeJSON,
        _1.includeDSON,
        tslib_1.__metadata("design:type", Object)
    ], RadixSerializableObject.prototype, "version", void 0);
    tslib_1.__decorate([
        _1.includeDSON,
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], RadixSerializableObject.prototype, "serializer", null);
    return RadixSerializableObject;
}());
exports.RadixSerializableObject = RadixSerializableObject;
