// prelude-ts/dist/src/Vector.js
"use strict";
exports.__esModule = true;
var Value_1 = require("./Value");
var Option_1 = require("./Option");
var HashMap_1 = require("./HashMap");
var HashSet_1 = require("./HashSet");
var Comparison_1 = require("./Comparison");
var SeqHelpers = require("./SeqHelpers");
var L = require("list");
/**
 * A general-purpose list class with all-around good performance.
 * quasi-O(1) (actually O(log32(n))) access, append, replace.
 * It's backed by a bit-mapped vector trie.
 * @param T the item type
 */
var Vector = /** @class */ (function () {
    /**
     * @hidden
     */
    // _contents will be undefined only if length===0
    function Vector(_list) {
        this._list = _list;
    }
    /**
     * The empty vector.
     * @param T the item type
     */
    Vector.empty = function () {
        return new Vector(L.empty());
    };
    /**
     * Build a vector from a series of items (any number, as parameters)
     * @param T the item type
     */
    Vector.of = function () {
        var data = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            data[_i] = arguments[_i];
        }
        return Vector.ofIterable(data);
    };
    /**
     * Build a vector from any iterable, which means also
     * an array for instance.
     * @param T the item type
     */
    Vector.ofIterable = function (elts) {
        return new Vector(L.from(elts));
    };
    /**
     * Curried predicate to find out whether the vector is empty.
     *
     *     LinkedList.of(Vector.of(1), Vector.empty<number>())
     *         .filter(Vector.isEmpty)
     *     => LinkedList.of(Vector.empty<number>())
     */
    Vector.isEmpty = function (v) {
        return v.isEmpty();
    };
    /**
     * Curried predicate to find out whether the vector is empty.
     *
     *     LinkedList.of(Vector.of(1), Vector.empty<number>())
     *         .filter(Vector.isNotEmpty)
     *     => LinkedList.of(Vector.of(1))
     */
    Vector.isNotEmpty = function (v) {
        return !v.isEmpty();
    };
    /**
     * Get the length of the collection.
     */
    Vector.prototype.length = function () {
        return L.length(this._list);
    };
    /**
     * true if the collection is empty, false otherwise.
     */
    Vector.prototype.isEmpty = function () {
        return L.length(this._list) === 0;
    };
    /**
     * Dual to the foldRight function. Build a collection from a seed.
     * Takes a starting element and a function.
     * It applies the function on the starting element; if the
     * function returns None, it stops building the list, if it
     * returns Some of a pair, it adds the first element to the result
     * and takes the second element as a seed to keep going.
     *
     *     Vector.unfoldRight(
     *          10, x=>Option.of(x)
     *              .filter(x => x!==0)
     *              .map<[number,number]>(x => [x,x-1]))
     *     => Vector.of(10, 9, 8, 7, 6, 5, 4, 3, 2, 1)
     */
    Vector.unfoldRight = function (seed, fn) {
        var nextVal = fn(seed);
        var r = L.empty();
        while (nextVal.isSome()) {
            r = L.append(nextVal.get()[0], r);
            nextVal = fn(nextVal.get()[1]);
        }
        return new Vector(r);
    };
    /**
     * Retrieve the element at index idx.
     * Returns an option because the collection may
     * contain less elements than the index.
     */
    Vector.prototype.get = function (index) {
        return Option_1.Option.of(L.nth(index, this._list));
    };
    /**
     * If the collection contains a single element,
     * return Some of its value, otherwise return None.
     */
    Vector.prototype.single = function () {
        return L.length(this._list) === 1 ?
            this.head() :
            Option_1.Option.none();
    };
    /**
     * Replace the value of element at the index you give.
     * Will throw if the index is out of bounds!
     */
    Vector.prototype.replace = function (index, val) {
        if (index >= this.length() || index < 0) {
            throw new Error('Vector.replace: index is out of range: ' + index);
        }
        return new Vector(L.update(index, val, this._list));
    };
    /**
     * Replace the first occurence (if any) of the element you give by
     * the new value you give.
     *
     *     Vector.of(1, 2, 3, 4, 2).replaceFirst(2, 5)
     *     => Vector.of(1, 5, 3, 4, 2)
     *
     */
    Vector.prototype.replaceFirst = function (element, newVal) {
        // it's a little annoying that areEqual will check whether the element
        // has an equals function for each element in the list, but then
        // what if the list allows null or undefined and the newVal is null or
        // undefined? With type erasure then I don't know what equality to use
        // on the next elements
        var index = L.findIndex(function (v) { return Comparison_1.areEqual(v, element); }, this._list);
        return (index >= 0)
            ? new Vector(L.update(index, newVal, this._list))
            : this;
    };
    /**
     * Replace all occurences of the element you give by
     * the new value you give.
     *
     *     Vector.of(1, 2, 3, 4, 2).replaceAll(2, 5)
     *     => Vector.of(1, 5, 3, 4, 5)
     *
     */
    Vector.prototype.replaceAll = function (element, newVal) {
        // if we're going to update many elements, then append in a loop
        // would give better perf (not copying multiple times the same slice).
        // if we won't update that many, update in a loop would give better perf...
        // assuming it's the latter case.
        var idx = 0;
        return this.foldLeft(this, function (sofar, cur) {
            var r = Comparison_1.areEqual(cur, element)
                ? new Vector(L.update(idx, newVal, sofar._list))
                : sofar;
            ++idx;
            return r;
        });
    };
    /**
     * Append an element at the end of the collection.
     */
    Vector.prototype.append = function (val) {
        return new Vector(L.append(val, this._list));
    };
    /**
     * Append multiple elements at the end of the collection.
     * Note that arrays are also iterables.
     */
    Vector.prototype.appendAll = function (elts) {
        if (elts._list && elts.replace) {
            // elts is a vector too
            return new Vector(L.concat(this._list, elts._list));
        }
        return new Vector(L.concat(this._list, L.from(elts)));
    };
    /**
     * Remove multiple elements from a vector
     *
     *     Vector.of(1,2,3,4,3,2,1).removeAll([2,4])
     *     => Vector.of(1,3,3,1)
     */
    Vector.prototype.removeAll = function (elts) {
        return SeqHelpers.removeAll(this, elts);
    };
    /**
     * Get the first value of the collection, if any.
     * returns Option.Some if the collection is not empty,
     * Option.None if it's empty.
     */
    Vector.prototype.head = function () {
        return this.get(0);
    };
    /**
     * Get the last value of the collection, if any.
     * returns Option.Some if the collection is not empty,
     * Option.None if it's empty.
     */
    Vector.prototype.last = function () {
        return Option_1.Option.of(L.last(this._list));
    };
    /**
     * Return a new vector containing all the elements in this
     * vector except the last one, or the empty vector if this
     * is the empty vector.
     *
     *     Vector.of(1,2,3).init()
     *     => Vector.of(1,2)
     */
    Vector.prototype.init = function () {
        return new Vector(L.pop(this._list));
    };
    /**
     * Returns a new collection, discarding the first elements
     * until one element fails the predicate. All elements
     * after that point are retained.
     */
    Vector.prototype.dropWhile = function (predicate) {
        return new Vector(L.dropWhile(predicate, this._list));
    };
    /**
     * Search for the first item matching the predicate you pass,
     * return Option.Some of that element if found,
     * Option.None otherwise.
     */
    Vector.prototype.find = function (predicate) {
        return Option_1.Option.of(L.find(predicate, this._list));
    };
    /**
     * Search for the last item matching the predicate you pass,
     * return Option.Some of that element if found,
     * Option.None otherwise.
     */
    Vector.prototype.findLast = function (predicate) {
        return Option_1.Option.of(L.findLast(predicate, this._list));
    };
    /**
     * Search for the first item matching the predicate you pass,
     * returning its index in the form of Option.Some if found,
     * Option.None otherwise.
     */
    Vector.prototype.findIndex = function (predicate) {
        return Option_1.Option.of(L.findIndex(predicate, this._list)).filter(function (i) { return i != -1; });
    };
    Vector.prototype.allMatch = function (predicate) {
        return L.every(predicate, this._list);
    };
    /**
     * Returns true if there the predicate returns true for any
     * element in the collection.
     */
    Vector.prototype.anyMatch = function (predicate) {
        return L.some(predicate, this._list);
    };
    Vector.prototype.partition = function (predicate) {
        return L.partition(predicate, this._list)
            .map(function (x) { return new Vector(x); });
    };
    /**
     * Returns true if the item is in the collection,
     * false otherwise.
     */
    Vector.prototype.contains = function (v) {
        return this.find(function (x) { return Comparison_1.areEqual(x, v); }).isSome();
    };
    /**
     * Group elements in the collection using a classifier function.
     * Elements are then organized in a map. The key is the value of
     * the classifier, and in value we get the list of elements
     * matching that value.
     *
     * also see [[Vector.arrangeBy]]
     */
    Vector.prototype.groupBy = function (classifier) {
        return this.foldLeft(HashMap_1.HashMap.empty(), function (acc, v) {
            return acc.putWithMerge(classifier(v), Vector.of(v), // !!! DOUBLE CHECK THIS
            function (v1, v2) { return v1.append(L.nth(0, v2._list)); });
        });
    };
    /**
     * Matches each element with a unique key that you extract from it.
     * If the same key is present twice, the function will return None.
     *
     * also see [[Vector.groupBy]]
     */
    Vector.prototype.arrangeBy = function (getKey) {
        return SeqHelpers.arrangeBy(this, getKey);
    };
    /**
     * Remove duplicate items; elements are mapped to keys, those
     * get compared.
     *
     *     Vector.of(1,1,2,3,2,3,1).distinctBy(x => x);
     *     => Vector.of(1,2,3)
     */
    Vector.prototype.distinctBy = function (keyExtractor) {
        return SeqHelpers.distinctBy(this, keyExtractor);
    };
    Vector.prototype[Symbol.iterator] = function () {
        return this._list[Symbol.iterator]();
    };
    /**
     * Call a function for element in the collection.
     */
    Vector.prototype.forEach = function (fun) {
        L.forEach(fun, this._list);
        return this;
    };
    /**
     * Return a new collection where each element was transformed
     * by the mapper function you give.
     */
    Vector.prototype.map = function (fun) {
        return new Vector(L.map(fun, this._list));
    };
    Vector.prototype.filter = function (fun) {
        return new Vector(L.filter(fun, this._list));
    };
    /**
     * Apply the mapper function on every element of this collection.
     * The mapper function returns an Option; if the Option is a Some,
     * the value it contains is added to the result Collection, if it's
     * a None, the value is discarded.
     *
     *     Vector.of(1,2,6).mapOption(x => x%2===0 ?
     *         Option.of(x+1) : Option.none<number>())
     *     => Vector.of(3, 7)
     */
    Vector.prototype.mapOption = function (mapper) {
        var vec = L.empty();
        for (var i = 0; i < this.length(); i++) {
            var v = mapper(L.nth(i, this._list));
            if (v.isSome()) {
                vec = L.append(v.get(), vec);
            }
        }
        return new Vector(vec);
    };
    /**
     * Calls the function you give for each item in the collection,
     * your function returns a collection, all the collections are
     * concatenated.
     * This is the monadic bind.
     */
    Vector.prototype.flatMap = function (mapper) {
        return new Vector(L.chain(function (x) { return mapper(x)._list; }, this._list));
    };
    /**
     * Reduces the collection to a single value using the
     * associative binary function you give. Since the function
     * is associative, order of application doesn't matter.
     *
     * Example:
     *
     *     Vector.of(1,2,3).fold(0, (a,b) => a + b);
     *     => 6
     */
    Vector.prototype.fold = function (zero, fn) {
        return this.foldLeft(zero, fn);
    };
    /**
     * Reduces the collection to a single value.
     * Left-associative.
     *
     * Example:
     *
     *     Vector.of("a", "b", "c").foldLeft("!", (xs,x) => x+xs);
     *     => "cba!"
     *
     * @param zero The initial value
     * @param fn A function taking the previous value and
     *           the current collection item, and returning
     *           an updated value.
     */
    Vector.prototype.foldLeft = function (zero, fn) {
        return L.foldl(fn, zero, this._list);
    };
    /**
     * Reduces the collection to a single value.
     * Right-associative.
     *
     * Example:
     *
     *     Vector.of("a", "b", "c").foldRight("!", (x,xs) => xs+x);
     *     => "!cba"
     *
     * @param zero The initial value
     * @param fn A function taking the current collection item and
     *           the previous value , and returning
     *           an updated value.
     */
    Vector.prototype.foldRight = function (zero, fn) {
        return L.foldr(fn, zero, this._list);
    };
    /**
     * Returns the index of the first occurence of the value you give, if present
     *
     *     Vector.of(1, 2, 3, 4, 3).indexOf(3)
     *     => Option.of(2)
     */
    Vector.prototype.indexOf = function (element) {
        return Option_1.Option.of(L.findIndex(function (v) { return Comparison_1.areEqual(v, element); }, this._list))
            .filter(function (i) { return i >= 0; });
    };
    /**
     * Randomly reorder the elements of the collection.
     */
    Vector.prototype.shuffle = function () {
        return Vector.ofIterable(SeqHelpers.shuffle(this.toArray()));
    };
    /**
     * Transform this value to another value type.
     * Enables fluent-style programming by chaining calls.
     */
    Vector.prototype.transform = function (converter) {
        return converter(this);
    };
    /**
     * Two objects are equal if they represent the same value,
     * regardless of whether they are the same object physically
     * in memory.
     */
    Vector.prototype.equals = function (other) {
        if (other === this) {
            return true;
        }
        if (!other || (!other._list) || (!L.isList(other._list))) {
            return false;
        }
        if (this.length() !== other.length())
            return false;
        for (var i = 0; i < this.length(); i++) {
            var myVal = L.nth(i, this._list);
            var hisVal = L.nth(i, other._list);
            if ((myVal === undefined) !== (hisVal === undefined)) {
                return false;
            }
            if (myVal === undefined || hisVal === undefined) {
                // they are both undefined, the || is for TS's flow analysis
                // so he realizes none of them is undefined after this.
                continue;
            }
            if (!Comparison_1.areEqual(myVal, hisVal)) {
                return false;
            }
        }
        return true;
    };
    /**
     * Get a number for that object. Two different values
     * may get the same number, but one value must always get
     * the same number. The formula can impact performance.
     */
    Vector.prototype.hashCode = function () {
        var hash = 1;
        for (var i = 0; i < this.length(); i++) {
            hash = 31 * hash + Comparison_1.getHashCode(L.nth(i, this._list));
        }
        return hash;
    };
    /**
     * Get a human-friendly string representation of that value.
     *
     * Also see [[Vector.mkString]]
     */
    Vector.prototype.toString = function () {
        var r = "Vector(";
        for (var i = 0; i < this.length(); i++) {
            if (i > 0) {
                r += ", ";
            }
            r += SeqHelpers.toStringHelper(L.nth(i, this._list));
        }
        return r + ")";
    };
    /**
     * Used by the node REPL to display values.
     * Most of the time should be the same as toString()
     */
    Vector.prototype[Value_1.inspect] = function () {
        return this.toString();
    };
    /**
     * Joins elements of the collection by a separator.
     * Example:
     *
     *     Vector.of(1,2,3).mkString(", ")
     *     => "1, 2, 3"
     */
    Vector.prototype.mkString = function (separator) {
        var r = "";
        for (var i = 0; i < this.length(); i++) {
            if (i > 0) {
                r += separator;
            }
            r += SeqHelpers.toStringHelper(L.nth(i, this._list), { quoteStrings: false });
        }
        return r;
    };
    /**
     * Returns a new collection with elements
     * sorted according to the comparator you give.
     *
     * also see [[Vector.sortOn]]
     */
    Vector.prototype.sortBy = function (compare) {
        return Vector.ofIterable(this.toArray().sort(compare));
    };
    /**
     * Give a function associating a number or a string with
     * elements from the collection, and the elements
     * are sorted according to that value.
     *
     *     Vector.of({a:3,b:"b"},{a:1,b:"test"},{a:2,b:"a"}).sortOn(elt=>elt.a)
     *     => Vector.of({a:1,b:"test"},{a:2,b:"a"},{a:3,b:"b"})
     *
     * You can also sort by multiple criteria, and request 'descending'
     * sorting:
     *
     *     Vector.of({a:1,b:"b"},{a:1,b:"test"},{a:2,b:"a"}).sortOn(elt=>elt.a,{desc:elt=>elt.b})
     *     => Vector.of({a:1,b:"test"},{a:1,b:"b"},{a:2,b:"a"})
     *
     * also see [[Vector.sortBy]]
     */
    Vector.prototype.sortOn = function () {
        var getKeys = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            getKeys[_i] = arguments[_i];
        }
        return SeqHelpers.sortOn(this, getKeys);
    };
    /**
     * Convert this collection to a map. You give a function which
     * for each element in the collection returns a pair. The
     * key of the pair will be used as a key in the map, the value,
     * as a value in the map. If several values get the same key,
     * entries will be lost.
     *
     *     Vector.of(1,2,3).toMap(x=>[x.toString(), x])
     *     => HashMap.of(["1",1], ["2",2], ["3",3])
     */
    Vector.prototype.toMap = function (converter) {
        return this.foldLeft(HashMap_1.HashMap.empty(), function (acc, cur) {
            var converted = converter(cur);
            return acc.put(converted[0], converted[1]);
        });
    };
    /**
     * Convert this collection to a set. Since the elements of the
     * Seq may not support equality, you must pass a function returning
     * a value supporting equality.
     *
     *     Vector.of(1,2,3,3,4).toSet(x=>x)
     *     => HashSet.of(1,2,3,4)
     */
    Vector.prototype.toSet = function (converter) {
        return this.foldLeft(HashSet_1.HashSet.empty(), function (acc, cur) {
            return acc.add(converter(cur));
        });
    };
    /**
     * Convert to array.
     */
    Vector.prototype.toArray = function () {
        return L.toArray(this._list);
    };
    ;
    /**
     * @hidden
     */
    Vector.prototype.hasTrueEquality = function () {
        return SeqHelpers.seqHasTrueEquality(this);
    };
    /**
     * Combine any number of iterables you give in as
     * parameters to produce a new collection which combines all,
     * in tuples. For instance:
     *
     *     Vector.zip(Vector.of(1,2,3), ["a","b","c"], LinkedList.of(8,9,10))
     *     => Vector.of([1,"a",8], [2,"b",9], [3,"c",10])
     *
     * The result collection will have the length of the shorter
     * of the input iterables. Extra elements will be discarded.
     *
     * Also see [the non-static version](#zip), which only combines two
     * collections.
     * @param A A is the type of the tuple that'll be generated
     *          (`[number,string,number]` for the code sample)
     */
    Vector.zip = function () {
        var iterables = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            iterables[_i] = arguments[_i];
        }
        var r = L.empty();
        var iterators = iterables.map(function (i) { return i[Symbol.iterator](); });
        var items = iterators.map(function (i) { return i.next(); });
        while (!items.some(function (item) { return item.done; })) {
            r = L.append(items.map(function (item) { return item.value; }), r);
            items = iterators.map(function (i) { return i.next(); });
        }
        return new Vector(r);
    };
    /**
     * Combine this collection with the collection you give in
     * parameter to produce a new collection which combines both,
     * in pairs. For instance:
     *
     *     Vector.of(1,2,3).zip(["a","b","c"])
     *     => Vector.of([1,"a"], [2,"b"], [3,"c"])
     *
     * The result collection will have the length of the shorter
     * of both collections. Extra elements will be discarded.
     *
     * Also see [[Vector.zip]] (static version which can more than two
     * iterables)
     */
    Vector.prototype.zip = function (other) {
        var r = L.empty();
        var thisIterator = this[Symbol.iterator]();
        var otherIterator = other[Symbol.iterator]();
        var thisCurItem = thisIterator.next();
        var otherCurItem = otherIterator.next();
        while (!thisCurItem.done && !otherCurItem.done) {
            r = L.append([thisCurItem.value, otherCurItem.value], r);
            thisCurItem = thisIterator.next();
            otherCurItem = otherIterator.next();
        }
        return new Vector(r);
    };
    /**
     * Reverse the collection. For instance:
     *
     *     Vector.of(1,2,3).reverse();
     *     => Vector.of(3,2,1)
     */
    Vector.prototype.reverse = function () {
        return new Vector(L.reverse(this._list));
    };
    /**
     * Combine this collection with the index of the elements
     * in it. Handy if you need the index when you map on
     * the collection for instance:
     *
     *     Vector.of("a","b").zipWithIndex().map(([v,idx]) => v+idx)
     *     => Vector.of("a0", "b1")
     */
    Vector.prototype.zipWithIndex = function () {
        return SeqHelpers.zipWithIndex(this);
    };
    /**
     * Returns a new collection, discarding the elements
     * after the first element which fails the predicate.
     */
    Vector.prototype.takeWhile = function (predicate) {
        return new Vector(L.takeWhile(predicate, this._list));
    };
    /**
     * Returns a new collection, discarding the elements
     * after the first element which fails the predicate,
     * but starting from the end of the collection.
     *
     *     Vector.of(1,2,3,4).takeRightWhile(x => x > 2)
     *     => Vector.of(3,4)
     */
    Vector.prototype.takeRightWhile = function (predicate) {
        return new Vector(L.takeLastWhile(predicate, this._list));
    };
    /**
     * Split the collection at a specific index.
     *
     *     Vector.of(1,2,3,4,5).splitAt(3)
     *     => [Vector.of(1,2,3), Vector.of(4,5)]
     */
    Vector.prototype.splitAt = function (index) {
        if (index < 0) {
            return [Vector.empty(), this];
        }
        return L.splitAt(index, this._list).map(function (x) { return new Vector(x); });
    };
    /**
     * Takes a predicate; returns a pair of collections.
     * The first one is the longest prefix of this collection
     * which satisfies the predicate, and the second collection
     * is the remainder of the collection.
     *
     *    Vector.of(1,2,3,4,5,6).span(x => x <3)
     *    => [Vector.of(1,2), Vector.of(3,4,5,6)]
     */
    Vector.prototype.span = function (predicate) {
        // could be potentially faster using splitAt.
        var first = this.takeWhile(predicate);
        return [first, this.drop(first.length())];
    };
    /**
     * Returns a new collection with the first
     * n elements discarded.
     * If the collection has less than n elements,
     * returns the empty collection.
     */
    Vector.prototype.drop = function (n) {
        return new Vector(L.drop(n, this._list));
    };
    /**
     * Return a new collection containing the first n
     * elements from this collection
     *
     *     Vector.of(1,2,3,4).take(2)
     *     => Vector.of(1,2)
     */
    Vector.prototype.take = function (n) {
        if (n < 0) {
            return Vector.empty();
        }
        return new Vector(L.take(n, this._list));
    };
    /**
     * Prepend an element at the beginning of the collection.
     */
    Vector.prototype.prepend = function (elt) {
        return new Vector(L.prepend(elt, this._list));
    };
    /**
     * Prepend multiple elements at the beginning of the collection.
     */
    Vector.prototype.prependAll = function (elts) {
        return Vector.ofIterable(elts).appendAll(this);
    };
    /**
     * Removes the first element matching the predicate
     * (use [[Seq.filter]] to remove all elements matching a predicate)
     */
    Vector.prototype.removeFirst = function (predicate) {
        var v1 = this.takeWhile(function (x) { return !predicate(x); });
        return v1.appendAll(this.drop(v1.length() + 1));
    };
    /**
     * Returns a new collection with the last
     * n elements discarded.
     * If the collection has less than n elements,
     * returns the empty collection.
     */
    Vector.prototype.dropRight = function (n) {
        if (n >= this.length()) {
            return Vector.empty();
        }
        return new Vector(L.dropLast(n, this._list));
    };
    /**
     * Returns a new collection, discarding the last elements
     * until one element fails the predicate. All elements
     * before that point are retained.
     */
    Vector.prototype.dropRightWhile = function (predicate) {
        var i = this.length() - 1;
        for (; i >= 0; i--) {
            if (!predicate(L.nth(i, this._list))) {
                return this.take(i + 1);
            }
        }
        return Vector.empty();
    };
    /**
     * Get all the elements in the collection but the first one.
     * If the collection is empty, return None.
     */
    Vector.prototype.tail = function () {
        if (this.isEmpty()) {
            return Option_1.Option.none();
        }
        return Option_1.Option.of(new Vector(L.tail(this._list)));
    };
    /**
     * Reduces the collection to a single value by repeatedly
     * calling the combine function.
     * No starting value. The order in which the elements are
     * passed to the combining function is undetermined.
     */
    Vector.prototype.reduce = function (combine) {
        return SeqHelpers.reduce(this, combine);
    };
    /**
     * Compare values in the collection and return the smallest element.
     * Returns Option.none if the collection is empty.
     *
     * also see [[Vector.minOn]]
     */
    Vector.prototype.minBy = function (compare) {
        return SeqHelpers.minBy(this, compare);
    };
    /**
     * Call the function you give for each value in the collection
     * and return the element for which the result was the smallest.
     * Returns Option.none if the collection is empty.
     *
     *     Vector.of({name:"Joe", age:12}, {name:"Paula", age:6}).minOn(x=>x.age)
     *     => Option.of({name:"Paula", age:6})
     *
     * also see [[Vector.minBy]]
     */
    Vector.prototype.minOn = function (getOrderable) {
        return SeqHelpers.minOn(this, getOrderable);
    };
    /**
     * Compare values in the collection and return the largest element.
     * Returns Option.none if the collection is empty.
     *
     * also see [[Vector.maxOn]]
     */
    Vector.prototype.maxBy = function (compare) {
        return SeqHelpers.maxBy(this, compare);
    };
    /**
     * Call the function you give for each value in the collection
     * and return the element for which the result was the largest.
     * Returns Option.none if the collection is empty.
     *
     *     Vector.of({name:"Joe", age:12}, {name:"Paula", age:6}).maxOn(x=>x.age)
     *     => Option.of({name:"Joe", age:12})
     *
     * also see [[Vector.maxBy]]
     */
    Vector.prototype.maxOn = function (getOrderable) {
        return SeqHelpers.maxOn(this, getOrderable);
    };
    /**
     * Call the function you give for each element in the collection
     * and sum all the numbers, return that sum.
     * Will return 0 if the collection is empty.
     *
     *     Vector.of(1,2,3).sumOn(x=>x)
     *     => 6
     */
    Vector.prototype.sumOn = function (getNumber) {
        return SeqHelpers.sumOn(this, getNumber);
    };
    /**
     * Slides a window of a specific size over the sequence.
     * Returns a lazy stream so memory use is not prohibitive.
     *
     *     Vector.of(1,2,3,4,5,6,7,8).sliding(3)
     *     => Stream.of(Vector.of(1,2,3), Vector.of(4,5,6), Vector.of(7,8))
     */
    Vector.prototype.sliding = function (count) {
        return SeqHelpers.sliding(this, count);
    };
    /**
     * Apply the function you give to all elements of the sequence
     * in turn, keeping the intermediate results and returning them
     * along with the final result in a list.
     * The last element of the result is the final cumulative result.
     *
     *     Vector.of(1,2,3).scanLeft(0, (soFar,cur)=>soFar+cur)
     *     => Vector.of(0,1,3,6)
     */
    Vector.prototype.scanLeft = function (init, fn) {
        return new Vector(L.scan(fn, init, this._list));
    };
    /**
     * Apply the function you give to all elements of the sequence
     * in turn, keeping the intermediate results and returning them
     * along with the final result in a list.
     * The first element of the result is the final cumulative result.
     *
     *     Vector.of(1,2,3).scanRight(0, (cur,soFar)=>soFar+cur)
     *     => Vector.of(6,5,3,0)
     */
    Vector.prototype.scanRight = function (init, fn) {
        var r = [];
        r.unshift(init);
        var cur = init;
        for (var i = this.length() - 1; i >= 0; i--) {
            cur = fn(L.nth(i, this._list), cur);
            r.unshift(cur);
        }
        return Vector.ofIterable(r);
    };
    return Vector;
}());
exports.Vector = Vector;
