// radixdlt/build/main/modules/universe/RadixUniverseConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var betanet_json_1 = tslib_1.__importDefault(require("./configs/betanet.json"));
var sunstone_json_1 = tslib_1.__importDefault(require("./configs/sunstone.json"));
var local_json_1 = tslib_1.__importDefault(require("./configs/local.json"));
var long_1 = tslib_1.__importDefault(require("long"));
var atommodel_1 = require("../atommodel");
var RadixUniverseConfig = /** @class */ (function () {
    function RadixUniverseConfig(rawJson) {
        this.rawJson = rawJson;
        this.magic = rawJson.magic;
        this.magicByte = long_1.default.fromNumber(this.magic).and(0xff).toNumber();
    }
    RadixUniverseConfig.prototype.initialize = function () {
        var obj = atommodel_1.RadixSerializer.fromJSON(this.rawJson);
        this.port = obj.port;
        this.name = obj.name;
        this.description = obj.description;
        this.type = obj.type;
        this.timestamp = obj.timestamp;
        this.creator = obj.creator;
        this.genesis = obj.genesis;
    };
    RadixUniverseConfig.prototype.getMagic = function () {
        return this.magic;
    };
    RadixUniverseConfig.prototype.getMagicByte = function () {
        return this.magicByte;
    };
    RadixUniverseConfig.SUNSTONE = new RadixUniverseConfig(sunstone_json_1.default);
    RadixUniverseConfig.LOCAL = new RadixUniverseConfig(local_json_1.default);
    RadixUniverseConfig.BETANET = new RadixUniverseConfig(betanet_json_1.default);
    return RadixUniverseConfig;
}());
exports.default = RadixUniverseConfig;
