// radixdlt/build/main/modules/universe/RadixNodeDiscoveryFromNodeFinder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var axios_1 = tslib_1.__importDefault(require("axios"));
var atommodel_1 = require("../atommodel");
var __1 = require("../..");
/**
 * Node discovery from the Radix bootstrap service
 */
var RadixNodeDiscoveryFromNodeFinder = /** @class */ (function () {
    /**
     * Creates an instance of radix node discovery from node finder.
     * @param bootstrapService The full address to the node finder service for the universe
     */
    function RadixNodeDiscoveryFromNodeFinder(bootstrapService, wsAddress, httpAddress) {
        this.bootstrapService = bootstrapService;
        this.wsAddress = wsAddress;
        this.httpAddress = httpAddress;
    }
    RadixNodeDiscoveryFromNodeFinder.prototype.loadNodes = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var bootstrapNodeIP, bootstrapNode, getPeersRequestData, nodeListResponse, nodeList;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, axios_1.default.get(this.bootstrapService)];
                    case 1:
                        bootstrapNodeIP = (_a.sent()).data;
                        if (bootstrapNodeIP.length < 2) {
                            throw new Error('Failed to get node list from bootstrap service');
                        }
                        bootstrapNode = this.httpAddress(bootstrapNodeIP, 0);
                        getPeersRequestData = {
                            id: 0,
                            method: 'Network.getLivePeers',
                            params: [],
                        };
                        return [4 /*yield*/, axios_1.default.post(bootstrapNode, getPeersRequestData)];
                    case 2:
                        nodeListResponse = _a.sent();
                        nodeList = atommodel_1.RadixSerializer.fromJSON(nodeListResponse.data.result);
                        if (nodeList.length === 0) {
                            throw new Error('Bootstrap node has no connections');
                        }
                        return [2 /*return*/, nodeList.map(function (peerInfo) {
                                return new __1.RadixNode(peerInfo.system, _this.wsAddress(peerInfo.host.ip, peerInfo.host.port), _this.httpAddress(peerInfo.host.ip, peerInfo.host.port));
                            })];
                }
            });
        });
    };
    return RadixNodeDiscoveryFromNodeFinder;
}());
exports.default = RadixNodeDiscoveryFromNodeFinder;
