// @radixdlt/primitives/dist/_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HRP = exports.hrpFullSuffixLength = exports.NetworkId = exports.Network = void 0;
var Network;
(function (Network) {
    Network["MAINNET"] = "mainnet";
    Network["STOKENET"] = "stokenet";
    Network["LOCALNET"] = "localnet";
    Network["MILESTONENET"] = "milestonenet";
    Network["RELEASENET"] = "releasenet";
    Network["RCNET"] = "rcnet";
    Network["TESTNET6"] = "testnet6";
    Network["SANDPITNET"] = "sandpitnet";
})(Network = exports.Network || (exports.Network = {}));
exports.NetworkId = {
    1: Network.MAINNET,
    2: Network.STOKENET,
    3: Network.RELEASENET,
    4: Network.RCNET,
    5: Network.MILESTONENET,
    6: Network.TESTNET6,
    7: Network.SANDPITNET,
    99: Network.LOCALNET,
};
exports.hrpFullSuffixLength = 3;
exports.HRP = {
    [Network.MAINNET]: {
        account: 'rdx',
        validator: 'rv',
        RRI_suffix: '_rr',
    },
    [Network.STOKENET]: {
        account: 'tdx',
        validator: 'tv',
        RRI_suffix: '_tr',
    },
    [Network.LOCALNET]: {
        account: 'ddx',
        validator: 'dv',
        RRI_suffix: '_dr',
    },
    [Network.RELEASENET]: {
        account: 'tdx3',
        validator: 'tv3',
        RRI_suffix: '_tr3',
    },
    [Network.RCNET]: {
        account: 'tdx4',
        validator: 'tv4',
        RRI_suffix: '_tr4',
    },
    [Network.MILESTONENET]: {
        account: 'tdx5',
        validator: 'tv5',
        RRI_suffix: '_tr5',
    },
    [Network.TESTNET6]: {
        account: 'tdx6',
        validator: 'tv6',
        RRI_suffix: '_tr6',
    },
    [Network.SANDPITNET]: {
        account: 'tdx7',
        validator: 'tv7',
        RRI_suffix: '_tr7',
    },
};
