// @radixdlt/crypto/dist/elliptic-curve/hd/bip39/_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LanguageT = exports.StrengthT = void 0;
var StrengthT;
(function (StrengthT) {
    /// Entropy of 128 bits
    StrengthT[StrengthT["WORD_COUNT_12"] = 128] = "WORD_COUNT_12";
    /// Entropy of 160 bits
    StrengthT[StrengthT["WORD_COUNT_15"] = 160] = "WORD_COUNT_15";
    /// Entropy of 192 bits
    StrengthT[StrengthT["WORD_COUNT_18"] = 192] = "WORD_COUNT_18";
    /// Entropy of 224 bits
    StrengthT[StrengthT["WORD_COUNT_21"] = 224] = "WORD_COUNT_21";
    /// Entropy of 256 bits
    StrengthT[StrengthT["WORD_COUNT_24"] = 256] = "WORD_COUNT_24";
})(StrengthT = exports.StrengthT || (exports.StrengthT = {}));
var LanguageT;
(function (LanguageT) {
    LanguageT[LanguageT["CZECH"] = 0] = "CZECH";
    LanguageT[LanguageT["CHINESE_SIMPLIFIED"] = 1] = "CHINESE_SIMPLIFIED";
    LanguageT[LanguageT["CHINESE_TRADITIONAL"] = 2] = "CHINESE_TRADITIONAL";
    LanguageT[LanguageT["KOREAN"] = 3] = "KOREAN";
    LanguageT[LanguageT["FRENCH"] = 4] = "FRENCH";
    LanguageT[LanguageT["ITALIAN"] = 5] = "ITALIAN";
    LanguageT[LanguageT["SPANISH"] = 6] = "SPANISH";
    LanguageT[LanguageT["JAPANESE"] = 7] = "JAPANESE";
    LanguageT[LanguageT["PORTUGUESE"] = 8] = "PORTUGUESE";
    LanguageT[LanguageT["ENGLISH"] = 9] = "ENGLISH";
})(LanguageT = exports.LanguageT || (exports.LanguageT = {}));
