// radixdlt/build/main/modules/account/RadixDataAccountSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var rxjs_1 = require("rxjs");
var typescript_map_1 = require("typescript-map");
var operators_1 = require("rxjs/operators");
var atommodel_1 = require("../atommodel");
var RadixDataAccountSystem = /** @class */ (function () {
    function RadixDataAccountSystem(address) {
        this.address = address;
        this.name = 'DATA';
        this.applicationDataSubject = new rxjs_1.Subject();
        this.applicationData = new typescript_map_1.TSMap();
    }
    RadixDataAccountSystem.prototype.processAtomUpdate = function (atomUpdate) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                if (!('decryptedData' in atomUpdate.processedData)) {
                    return [2 /*return*/];
                }
                if (atomUpdate.action === 'STORE') {
                    this.processStoreAtom(atomUpdate);
                }
                else if (atomUpdate.action === 'DELETE') {
                    this.processDeleteAtom(atomUpdate);
                }
                return [2 /*return*/];
            });
        });
    };
    RadixDataAccountSystem.prototype.processStoreAtom = function (atomUpdate) {
        var atom = atomUpdate.atom;
        var aid = atom.getAidString();
        var signatures = atom.signatures;
        var applicationId = atomUpdate.processedData.decryptedData.application;
        // Skip existing atoms
        if (this.applicationData.has(applicationId) &&
            this.applicationData.get(applicationId).has(aid)) {
            return;
        }
        var applicationData = {
            aid: aid,
            payload: atomUpdate.processedData.decryptedData,
            timestamp: atom.getTimestamp(),
            signatures: signatures,
        };
        var applicationDataUpdate = {
            action: 'STORE',
            aid: aid,
            applicationId: applicationId,
            data: applicationData,
            signatures: signatures,
        };
        if (!this.applicationData.has(applicationId)) {
            this.applicationData.set(applicationId, new typescript_map_1.TSMap());
        }
        this.applicationData.get(applicationId).set(aid, applicationData);
        this.applicationDataSubject.next(applicationDataUpdate);
    };
    RadixDataAccountSystem.prototype.processDeleteAtom = function (atomUpdate) {
        var atom = atomUpdate.atom;
        var aid = atom.getAidString();
        var signatures = atom.signatures;
        var applicationId = atomUpdate.processedData.decryptedData.application;
        // Skip nonexisting atoms
        if (!this.applicationData.has(applicationId) ||
            !this.applicationData.get(applicationId).has(aid)) {
            return;
        }
        var applicationData = this.applicationData.get(applicationId).get(aid);
        var applicationDataUpdate = {
            action: 'DELETE',
            aid: aid,
            applicationId: applicationId,
            data: applicationData,
            signatures: signatures,
        };
        this.applicationData.get(applicationId).delete(aid);
        this.applicationDataSubject.next(applicationDataUpdate);
    };
    /**
     * Gets application data messages by application id and optionally by signer
     *
     * @param applicationId - Application id of the payload atoms
     * @param [addresses] - List of account addreses to filter application messages by signer
     * @returns An observable subscribed to old and new application messages that met the filter requirements
     */
    RadixDataAccountSystem.prototype.getApplicationData = function (applicationId, addresses) {
        var _this = this;
        // Pre-calculate signatureIds
        var signatureIds = !addresses ? undefined : addresses.map(function (a) { return atommodel_1.RadixAddress.fromAddress(a).getUID().toString(); });
        return rxjs_1.Observable.create(function (observer) {
            var e_1, _a;
            // Send all old data
            if (_this.applicationData.has(applicationId)) {
                var _loop_1 = function (applicationData) {
                    if (!signatureIds
                        || signatureIds.length === 0
                        || signatureIds.some(function (s) { return Object.keys(applicationData.signatures).includes(s); })) {
                        var applicationDataUpdate = {
                            action: 'STORE',
                            aid: applicationData.aid,
                            applicationId: applicationId,
                            data: applicationData,
                            signatures: applicationData.signatures,
                        };
                        observer.next(applicationDataUpdate);
                    }
                };
                try {
                    for (var _b = tslib_1.__values(_this.applicationData
                        .get(applicationId)
                        .values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var applicationData = _c.value;
                        _loop_1(applicationData);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            // Subscribe for new ones
            _this.applicationDataSubject
                .pipe(operators_1.filter(function (update) {
                return update.applicationId === applicationId
                    && (!signatureIds
                        || signatureIds.length === 0
                        || signatureIds.some(function (s) { return Object.keys(update.signatures).includes(s); }));
            }))
                .subscribe(observer);
        });
    };
    return RadixDataAccountSystem;
}());
exports.default = RadixDataAccountSystem;
