// radixdlt/build/main/modules/account/RadixDecryptionAccountSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var atommodel_1 = require("../atommodel");
var RadixDecryptionState;
(function (RadixDecryptionState) {
    RadixDecryptionState["DECRYPTED"] = "DECRYPTED";
    RadixDecryptionState["NOT_ENCRYPTED"] = "NOT_ENCRYPTED";
    RadixDecryptionState["CANNOT_DECRYPT"] = "CANNOT_DECRYPT";
})(RadixDecryptionState = exports.RadixDecryptionState || (exports.RadixDecryptionState = {}));
var RadixDecryptionAccountSystem = /** @class */ (function () {
    function RadixDecryptionAccountSystem(decryptionProvider) {
        this.name = 'DECRYPTION';
        if (decryptionProvider) {
            this.decryptionProvider = decryptionProvider;
        }
    }
    RadixDecryptionAccountSystem.prototype.processAtomUpdate = function (atomUpdate) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var atom, messageParticles, dataParticle, encryptorParticle, decryptedData, protectors, payload, decrypted, error_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        atom = atomUpdate.atom;
                        messageParticles = atom.getParticlesOfType(atommodel_1.RadixMessageParticle);
                        dataParticle = messageParticles.find(function (p) {
                            return p.getMetaData('application') !== 'encryptor';
                        });
                        encryptorParticle = messageParticles.find(function (p) {
                            return p.getMetaData('application') === 'encryptor';
                        });
                        if (!(dataParticle)) {
                            return [2 /*return*/];
                        }
                        if (!encryptorParticle) return [3 /*break*/, 5];
                        protectors = encryptorParticle.getData().asJSON().map(function (p) { return Buffer.from(p, 'base64'); });
                        payload = dataParticle.getData().bytes;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.decryptionProvider.decryptECIESPayloadWithProtectors(protectors, payload)];
                    case 2:
                        decrypted = _a.sent();
                        decryptedData = {
                            data: decrypted.toString(),
                            decryptionState: RadixDecryptionState.DECRYPTED,
                            application: dataParticle.getMetaData('application'),
                            from: dataParticle.from,
                            to: dataParticle.getAddresses(),
                        };
                        return [3 /*break*/, 4];
                    case 3:
                        error_1 = _a.sent();
                        decryptedData = {
                            data: dataParticle.getData().toString(),
                            decryptionState: RadixDecryptionState.CANNOT_DECRYPT,
                            application: dataParticle.getMetaData('application'),
                            from: dataParticle.from,
                            to: dataParticle.getAddresses(),
                        };
                        return [3 /*break*/, 4];
                    case 4: return [3 /*break*/, 6];
                    case 5:
                        decryptedData = {
                            data: dataParticle.getData().bytes.toString(),
                            decryptionState: RadixDecryptionState.NOT_ENCRYPTED,
                            application: dataParticle.getMetaData('application'),
                            from: dataParticle.from,
                            to: dataParticle.getAddresses(),
                        };
                        _a.label = 6;
                    case 6:
                        atomUpdate.processedData.decryptedData = decryptedData;
                        return [2 /*return*/];
                }
            });
        });
    };
    return RadixDecryptionAccountSystem;
}());
exports.RadixDecryptionAccountSystem = RadixDecryptionAccountSystem;
