// @radixdlt/util/dist/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogLevel = exports.restoreDefaultLogLevel = exports.log = void 0;
const tslib_1 = require("tslib");
const loglevel_1 = tslib_1.__importDefault(require("loglevel"));
exports.log = loglevel_1.default;
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const loglevel_plugin_prefix_1 = tslib_1.__importDefault(require("loglevel-plugin-prefix"));
var LogLevel;
(function (LogLevel) {
    LogLevel["SILENT"] = "silent";
    LogLevel["TRACE"] = "trace";
    LogLevel["DEBUG"] = "debug";
    LogLevel["INFO"] = "info";
    LogLevel["WARN"] = "warn";
    LogLevel["ERROR"] = "error";
})(LogLevel || (LogLevel = {}));
exports.LogLevel = LogLevel;
const defaultLogLevel = LogLevel.WARN;
const restoreDefaultLogLevel = () => {
    loglevel_1.default.setLevel(defaultLogLevel);
};
exports.restoreDefaultLogLevel = restoreDefaultLogLevel;
restoreDefaultLogLevel();
const logDecorations = {
    trace: {
        color: chalk_1.default.italic.cyan,
        emoji: 'ðŸ’œ',
    },
    debug: {
        color: chalk_1.default.italic.cyan,
        emoji: 'ðŸ’š',
    },
    info: {
        color: chalk_1.default.blue,
        emoji: 'ðŸ’™',
    },
    warn: {
        color: chalk_1.default.yellow,
        emoji: 'ðŸ’›',
    },
    error: {
        color: chalk_1.default.red,
        emoji: 'â¤ï¸',
    },
};
loglevel_plugin_prefix_1.default.reg(loglevel_1.default);
loglevel_plugin_prefix_1.default.apply(loglevel_1.default, {
    format: (level, name, timestamp) => `${chalk_1.default.gray(`[${timestamp.toString()}]`)} ${logDecorations[level.toLowerCase()]
        .emoji} ${logDecorations[level.toLowerCase()].color(level)}`,
});
