// @radixdlt/application/dist/api/json-rpc/_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiMethod = void 0;
var ApiMethod;
(function (ApiMethod) {
    ApiMethod["NETWORK_ID"] = "network.get_id";
    ApiMethod["TOKEN_BALANCES"] = "account.get_balances";
    ApiMethod["TRANSACTION_HISTORY"] = "account.get_transaction_history";
    ApiMethod["STAKES"] = "account.get_stake_positions";
    ApiMethod["UNSTAKES"] = "account.get_unstake_positions";
    ApiMethod["TX_STATUS"] = "transactions.get_transaction_status";
    ApiMethod["NETWORK_TX_THROUGHPUT"] = "network.get_throughput";
    ApiMethod["NETWORK_TX_DEMAND"] = "network.get_demand";
    ApiMethod["VALIDATORS"] = "validators.get_next_epoch_set";
    ApiMethod["LOOKUP_TX"] = "transactions.lookup_transaction";
    ApiMethod["LOOKUP_VALIDATOR"] = "validators.lookup_validator";
    ApiMethod["NATIVE_TOKEN"] = "tokens.get_native_token";
    ApiMethod["TOKEN_INFO"] = "tokens.get_info";
    ApiMethod["BUILD_TX_FROM_INTENT"] = "construction.build_transaction";
    ApiMethod["SUBMIT_TX"] = "construction.submit_transaction";
    ApiMethod["FINALIZE_TX"] = "construction.finalize_transaction";
})(ApiMethod = exports.ApiMethod || (exports.ApiMethod = {}));
