// radixdlt/build/main/modules/crypto/RadixKeyStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var atommodel_1 = require("../atommodel");
var crypto_1 = tslib_1.__importDefault(require("crypto"));
var RadixUtil_1 = require("../common/RadixUtil");
var RadixKeyStore = /** @class */ (function () {
    function RadixKeyStore() {
    }
    /**
     * Encrypt a private key
     * @param address
     * @param password
     * @returns
     */
    RadixKeyStore.encryptKey = function (address, password) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var privateKey = address.keyPair.getPrivate('hex');
            // Derrive key
            var salt = crypto_1.default.randomBytes(32).toString('hex');
            var iterations = 100000;
            var keylen = 32;
            var digest = 'sha512';
            crypto_1.default.pbkdf2(password, salt, iterations, keylen, digest, function (error, derivedKey) {
                if (error) {
                    reject(error);
                }
                // Encrypt private keys with derrived key
                var algorithm = 'aes-256-ctr';
                var iv = crypto_1.default.randomBytes(16);
                var cipher = crypto_1.default.createCipheriv(algorithm, derivedKey, iv);
                var ciphertext = Buffer.concat([
                    cipher.update(privateKey),
                    cipher.final(),
                ]);
                // Compute MAC
                var mac = _this.calculateMac(derivedKey, ciphertext);
                var fileContents = {
                    crypto: {
                        cipher: algorithm,
                        cipherparams: {
                            iv: iv.toString('hex'),
                        },
                        ciphertext: ciphertext.toString('hex'),
                        pbkdfparams: {
                            iterations: iterations,
                            keylen: keylen,
                            digest: digest,
                            salt: salt,
                        },
                        mac: mac.toString('hex'),
                    },
                    id: address.getUID().toString(),
                };
                resolve(fileContents);
            });
        });
    };
    /**
     * Decrypts an encrypted private key
     * @param fileContents
     * @param password
     * @returns key
     */
    RadixKeyStore.decryptKey = function (fileContents, password) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            // Derrive key
            var salt = fileContents.crypto.pbkdfparams.salt;
            var iterations = fileContents.crypto.pbkdfparams.iterations;
            var keylen = fileContents.crypto.pbkdfparams.keylen;
            var digest = fileContents.crypto.pbkdfparams.digest;
            crypto_1.default.pbkdf2(password, salt, iterations, keylen, digest, function (error, derivedKey) {
                if (error) {
                    return reject(error);
                }
                // Decrypt ciphertext
                var algorithm = fileContents.crypto.cipher;
                var iv = Buffer.from(fileContents.crypto.cipherparams.iv, 'hex');
                var ciphertext = Buffer.from(fileContents.crypto.ciphertext, 'hex');
                // Check MAC
                var MAC = Buffer.from(fileContents.crypto.mac, 'hex');
                var computedMAC = _this.calculateMac(derivedKey, ciphertext);
                if (!computedMAC.equals(MAC)) {
                    return reject('MAC mismatch');
                }
                var decipher = crypto_1.default.createDecipheriv(algorithm, derivedKey, iv);
                var privateKey = Buffer.concat([
                    decipher.update(ciphertext),
                    decipher.final(),
                ]).toString();
                // Create wallet
                var keyPair = atommodel_1.RadixAddress.fromPrivate(privateKey);
                return resolve(keyPair);
            });
        });
    };
    RadixKeyStore.calculateMac = function (derivedKey, ciphertext) {
        var dataToMac = Buffer.concat([derivedKey, ciphertext]);
        return RadixUtil_1.radixHash(dataToMac);
    };
    return RadixKeyStore;
}());
exports.default = RadixKeyStore;
