// radixdlt/build/main/modules/atommodel/atom/RadixAtom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var __1 = require("..");
var StringifySet_1 = require("../../common/StringifySet");
var RadixAtom = /** @class */ (function (_super) {
    tslib_1.__extends(RadixAtom, _super);
    function RadixAtom() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.particleGroups = [];
        _this.metaData = {};
        return _this;
    }
    RadixAtom_1 = RadixAtom;
    RadixAtom.prototype.getParticles = function () {
        var e_1, _a, e_2, _b;
        var particles = [];
        try {
            for (var _c = tslib_1.__values(this.particleGroups), _d = _c.next(); !_d.done; _d = _c.next()) {
                var particleGroup = _d.value;
                try {
                    for (var _e = (e_2 = void 0, tslib_1.__values(particleGroup.particles)), _f = _e.next(); !_f.done; _f = _e.next()) {
                        var particle = _f.value;
                        particles.push(particle);
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return particles;
    };
    RadixAtom.prototype.getAddresses = function () {
        var e_3, _a, e_4, _b, e_5, _c;
        var addressSet = new StringifySet_1.StringifySet();
        try {
            for (var _d = tslib_1.__values(this.particleGroups), _e = _d.next(); !_e.done; _e = _d.next()) {
                var particleGroup = _e.value;
                try {
                    for (var _f = (e_4 = void 0, tslib_1.__values(particleGroup.particles)), _g = _f.next(); !_g.done; _g = _f.next()) {
                        var particle = _g.value;
                        var addresses = particle.particle.getAddresses();
                        try {
                            for (var addresses_1 = (e_5 = void 0, tslib_1.__values(addresses)), addresses_1_1 = addresses_1.next(); !addresses_1_1.done; addresses_1_1 = addresses_1.next()) {
                                var address = addresses_1_1.value;
                                addressSet.add(address);
                            }
                        }
                        catch (e_5_1) { e_5 = { error: e_5_1 }; }
                        finally {
                            try {
                                if (addresses_1_1 && !addresses_1_1.done && (_c = addresses_1.return)) _c.call(addresses_1);
                            }
                            finally { if (e_5) throw e_5.error; }
                        }
                    }
                }
                catch (e_4_1) { e_4 = { error: e_4_1 }; }
                finally {
                    try {
                        if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                    }
                    finally { if (e_4) throw e_4.error; }
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return addressSet.values();
    };
    RadixAtom.prototype.getShards = function () {
        var e_6, _a, e_7, _b, e_8, _c;
        var shardSet = new StringifySet_1.StringifySet();
        try {
            for (var _d = tslib_1.__values(this.particleGroups), _e = _d.next(); !_e.done; _e = _d.next()) {
                var particleGroup = _e.value;
                try {
                    for (var _f = (e_7 = void 0, tslib_1.__values(particleGroup.particles)), _g = _f.next(); !_g.done; _g = _f.next()) {
                        var particle = _g.value;
                        var addresses = particle.particle.getAddresses();
                        try {
                            for (var addresses_2 = (e_8 = void 0, tslib_1.__values(addresses)), addresses_2_1 = addresses_2.next(); !addresses_2_1.done; addresses_2_1 = addresses_2.next()) {
                                var address = addresses_2_1.value;
                                var shard = address.getShard();
                                shardSet.add(shard);
                            }
                        }
                        catch (e_8_1) { e_8 = { error: e_8_1 }; }
                        finally {
                            try {
                                if (addresses_2_1 && !addresses_2_1.done && (_c = addresses_2.return)) _c.call(addresses_2);
                            }
                            finally { if (e_8) throw e_8.error; }
                        }
                    }
                }
                catch (e_7_1) { e_7 = { error: e_7_1 }; }
                finally {
                    try {
                        if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                    }
                    finally { if (e_7) throw e_7.error; }
                }
            }
        }
        catch (e_6_1) { e_6 = { error: e_6_1 }; }
        finally {
            try {
                if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
            }
            finally { if (e_6) throw e_6.error; }
        }
        return shardSet.values();
    };
    RadixAtom.prototype.getTimestamp = function () {
        var timestamp = parseInt(this.metaData[RadixAtom_1.METADATA_TIMESTAMP_KEY], 10);
        if (Number.isNaN(timestamp)) {
            throw new Error('Timestamp is not set or not a valid number');
        }
        else {
            return timestamp;
        }
    };
    RadixAtom.prototype.setTimestamp = function (timestamp) {
        this.metaData[RadixAtom_1.METADATA_TIMESTAMP_KEY] = '' + timestamp;
    };
    RadixAtom.prototype.setPowNonce = function (nonce) {
        this.metaData[RadixAtom_1.METADATA_POW_NONCE_KEY] = nonce.toString();
    };
    RadixAtom.prototype.clearPowNonce = function () {
        delete this.metaData[RadixAtom_1.METADATA_POW_NONCE_KEY];
    };
    RadixAtom.prototype.getSpunParticlesOfType = function () {
        var types = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            types[_i] = arguments[_i];
        }
        return this.getParticles()
            .filter(function (s) {
            var e_9, _a;
            try {
                for (var types_1 = tslib_1.__values(types), types_1_1 = types_1.next(); !types_1_1.done; types_1_1 = types_1.next()) {
                    var type = types_1_1.value;
                    if (s.particle instanceof type) {
                        return true;
                    }
                }
            }
            catch (e_9_1) { e_9 = { error: e_9_1 }; }
            finally {
                try {
                    if (types_1_1 && !types_1_1.done && (_a = types_1.return)) _a.call(types_1);
                }
                finally { if (e_9) throw e_9.error; }
            }
        });
    };
    RadixAtom.prototype.getParticlesOfSpin = function (spin) {
        return this.getParticles()
            .filter(function (spunParticle) { return spunParticle.spin === spin; })
            .map(function (spunParticle) { return spunParticle.particle; });
    };
    RadixAtom.prototype.getParticlesOfType = function (type, spin) {
        var particles = this.getParticles()
            .filter(function (spunParticle) { return spunParticle.particle instanceof type; });
        if (spin) {
            particles = particles.filter(function (spunParticle) { return spunParticle.spin === spin; });
        }
        return particles.map(function (spunParticle) { return spunParticle.particle; });
    };
    RadixAtom.prototype.getFirstParticleOfType = function (type) {
        return this.getParticles()
            .find(function (spunParticle) { return spunParticle.particle instanceof type; }).particle;
    };
    RadixAtom.prototype.containsParticle = function () {
        var e_10, _a, e_11, _b;
        var types = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            types[_i] = arguments[_i];
        }
        try {
            for (var _c = tslib_1.__values(this.getParticles()), _d = _c.next(); !_d.done; _d = _c.next()) {
                var spunParticle = _d.value;
                try {
                    for (var types_2 = (e_11 = void 0, tslib_1.__values(types)), types_2_1 = types_2.next(); !types_2_1.done; types_2_1 = types_2.next()) {
                        var type = types_2_1.value;
                        if (spunParticle.particle instanceof type) {
                            return true;
                        }
                    }
                }
                catch (e_11_1) { e_11 = { error: e_11_1 }; }
                finally {
                    try {
                        if (types_2_1 && !types_2_1.done && (_b = types_2.return)) _b.call(types_2);
                    }
                    finally { if (e_11) throw e_11.error; }
                }
            }
        }
        catch (e_10_1) { e_10 = { error: e_10_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_10) throw e_10.error; }
        }
        return false;
    };
    RadixAtom.prototype.getParticleGroups = function () {
        return this.particleGroups;
    };
    RadixAtom.prototype.getAid = function () {
        return __1.RadixAID.from(this.getHash(), this.getShards());
    };
    RadixAtom.prototype.getAidString = function () {
        return this.getAid().toString();
    };
    var RadixAtom_1;
    RadixAtom.METADATA_TIMESTAMP_KEY = 'timestamp';
    RadixAtom.METADATA_POW_NONCE_KEY = 'powNonce';
    tslib_1.__decorate([
        __1.includeJSON,
        __1.includeDSON,
        tslib_1.__metadata("design:type", Array)
    ], RadixAtom.prototype, "particleGroups", void 0);
    tslib_1.__decorate([
        __1.includeJSON,
        tslib_1.__metadata("design:type", Object)
    ], RadixAtom.prototype, "signatures", void 0);
    tslib_1.__decorate([
        __1.includeDSON,
        __1.includeJSON,
        tslib_1.__metadata("design:type", Object)
    ], RadixAtom.prototype, "metaData", void 0);
    RadixAtom = RadixAtom_1 = tslib_1.__decorate([
        __1.RadixSerializer.registerClass('radix.atom')
    ], RadixAtom);
    return RadixAtom;
}(__1.RadixSerializableObject));
exports.RadixAtom = RadixAtom;
