// radixdlt/build/module/modules/atommodel/primitives/RadixHash.js
import * as tslib_1 from "tslib";
import { RadixSerializer } from '..';
const id = ':hsh:';
let RadixHash = class RadixHash {
    constructor(bytes) {
        if (bytes.length !== 64) {
            throw new Error('Hash must be 64 bytes');
        }
        if (typeof bytes === 'string') {
            this.bytes = Buffer.from(bytes, 'hex');
        }
        else {
            this.bytes = Buffer.from(bytes);
        }
    }
    static fromJSON(data) {
        return new this(data);
    }
    toJSON() {
        return `${id}${this.bytes.toString('hex')}`;
    }
    toDSON() {
        return RadixSerializer.toDSON(this);
    }
    encodeCBOR(encoder) {
        const output = Buffer.alloc(this.bytes.length + 1);
        output.writeInt8(0x03, 0);
        this.bytes.copy(output, 1);
        return encoder.pushAny(output);
    }
};
RadixHash = tslib_1.__decorate([
    RadixSerializer.registerPrimitive(id),
    tslib_1.__metadata("design:paramtypes", [Object])
], RadixHash);
export { RadixHash };
