// radixdlt/build/module/modules/atommodel/primitives/RadixAID.js
var RadixAID_1;
import * as tslib_1 from "tslib";
import long from 'long';
import { RadixSerializer } from '..';
const id = ':aid:';
let RadixAID = RadixAID_1 = class RadixAID {
    constructor(bytes) {
        if (bytes.length !== RadixAID_1.BYTES) {
            throw new Error(`Bytest lenght must be ${RadixAID_1.BYTES} but is ${bytes.length}`);
        }
        this.bytes = Buffer.from(bytes);
    }
    static from(hash, shards) {
        if (shards.length === 0) {
            throw new Error('Shards array cannot be empty');
        }
        const shardIndex = hash[0] % shards.length;
        const selectedShard = shards
            .map(s => s.toUnsigned())
            .sort((a, b) => a.compare(b))
            .map(s => s.toSigned())[shardIndex];
        const bytes = Buffer.alloc(this.BYTES);
        hash.copy(bytes, 0, 0, this.HASH_BYTES);
        Buffer.from(selectedShard.toBytes()).copy(bytes, this.HASH_BYTES);
        return new this(bytes);
    }
    static fromJSON(data) {
        return new this(Buffer.from(data, 'hex'));
    }
    toJSON() {
        return `${id}${this.bytes.toString('hex')}`;
    }
    toDSON() {
        return RadixSerializer.toDSON(this);
    }
    encodeCBOR(encoder) {
        const output = Buffer.alloc(this.bytes.length + 1);
        output.writeInt8(0x08, 0);
        this.bytes.copy(output, 1);
        return encoder.pushAny(output);
    }
    equals(aid) {
        return this.bytes.compare(aid.bytes) === 0;
    }
    toString() {
        return this.bytes.toString('hex');
    }
    getShard() {
        return long.fromBytes(Array.from(this.bytes.subarray(RadixAID_1.HASH_BYTES)));
    }
};
RadixAID.BYTES = 32;
RadixAID.HASH_BYTES = 24;
RadixAID.SHARD_BYTES = 8;
RadixAID = RadixAID_1 = tslib_1.__decorate([
    RadixSerializer.registerPrimitive(id),
    tslib_1.__metadata("design:paramtypes", [Buffer])
], RadixAID);
export { RadixAID };
