// radixdlt/build/module/modules/atommodel/primitives/RadixBytes.js
import * as tslib_1 from "tslib";
import { RadixSerializer } from '..';
let RadixBytes = class RadixBytes {
    constructor(bytes) {
        this.bytes = Buffer.from(bytes);
    }
    static fromEncoded(encoded) {
        return new this(Buffer.from(encoded, 'base64'));
    }
    static fromJSON(bytes) {
        return this.fromEncoded(bytes);
    }
    toJSON() {
        return `:byt:${this.bytes.toString('base64')}`;
    }
    toDSON() {
        return RadixSerializer.toDSON(this);
    }
    encodeCBOR(encoder) {
        const output = Buffer.alloc(this.bytes.length + 1);
        output.writeInt8(0x01, 0);
        this.bytes.copy(output, 1);
        return encoder.pushAny(output);
    }
    toString() {
        return this.bytes.toString();
    }
    asJSON() {
        return JSON.parse(this.bytes.toString('utf8'));
    }
};
RadixBytes = tslib_1.__decorate([
    RadixSerializer.registerPrimitive(':byt:'),
    tslib_1.__metadata("design:paramtypes", [Object])
], RadixBytes);
export { RadixBytes };
