// radixdlt/build/module/modules/atommodel/primitives/RadixEUID.js
import * as tslib_1 from "tslib";
import { RadixSerializer } from '..';
import long from 'long';
const id = ':uid:';
let RadixEUID = class RadixEUID {
    constructor(value) {
        if (typeof value === 'number') {
            this.bytes = Buffer.alloc(16);
            this.bytes.writeUInt32BE(value, 12);
        }
        else if (Buffer.isBuffer(value) || Array.isArray(value)) {
            if (value.length === 0) {
                throw new Error('EUID must not be 0 bytes');
            }
            this.bytes = Buffer.from(value);
        }
        else if (typeof value === 'string') {
            this.bytes = Buffer.from(value, 'hex');
        }
        else {
            throw new Error('Unsupported EUID value');
        }
        this.shard = long.fromBytes([...this.bytes.slice(0, this.bytes.length - 8)]);
    }
    static fromJSON(data) {
        return new this(data);
    }
    toJSON() {
        return `${id}${this.bytes.toString('hex')}`;
    }
    toDSON() {
        return RadixSerializer.toDSON(this);
    }
    encodeCBOR(encoder) {
        const output = Buffer.alloc(this.bytes.length + 1);
        output.writeInt8(0x02, 0);
        this.bytes.copy(output, 1);
        return encoder.pushAny(output);
    }
    equals(euid) {
        return this.bytes.compare(euid.bytes) === 0;
    }
    toString() {
        return this.bytes.toString('hex');
    }
    getShard() {
        return this.shard;
    }
};
RadixEUID = tslib_1.__decorate([
    RadixSerializer.registerPrimitive(id),
    tslib_1.__metadata("design:paramtypes", [Object])
], RadixEUID);
export { RadixEUID };
