// radixdlt/build/module/modules/atommodel/primitives/RRI.js
import * as tslib_1 from "tslib";
import { RadixSerializer, RadixAddress, } from '..';
const id = ':rri:';
let RRI = class RRI {
    constructor(address, name) {
        this.address = address;
        this.name = name;
    }
    static fromJSON(uri) {
        return this.fromString(uri);
    }
    static fromString(uri) {
        const parts = uri.split('/');
        if (parts.length !== 3) {
            throw new Error('RRI must be of the format /:address/:unique');
        }
        return new this(RadixAddress.fromAddress(parts[1]), parts[2]);
    }
    toJSON() {
        return `${id}${this.toString()}`;
    }
    toString() {
        return `/${this.address.toString()}/${this.name}`;
    }
    toDSON() {
        return RadixSerializer.toDSON(this);
    }
    encodeCBOR(encoder) {
        const s = Buffer.from(this.toString(), 'utf8');
        const output = Buffer.alloc(s.length + 1);
        output.writeInt8(0x06, 0);
        s.copy(output, 1);
        return encoder.pushAny(output);
    }
    equals(rri) {
        return this.address.equals(rri.address) && this.name === rri.name;
    }
    getAddress() {
        return this.address;
    }
    getName() {
        return this.name;
    }
};
RRI = tslib_1.__decorate([
    RadixSerializer.registerPrimitive(id),
    tslib_1.__metadata("design:paramtypes", [RadixAddress, String])
], RRI);
export { RRI };
